---
title: "Firre's expression profile"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(ggrepel)
library(ggbeeswarm)
source("/scratch/Shares/rinn/JR/Firre_time/util/_plot_theme.R")
source("/scratch/Shares/rinn/JR/Firre_time/util/_util.R")
```

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData")
```

We would like to check that Firre's expression is actually induced by the addition of
doxycycline.

```{r Firre induction in WT & KO background at time point 0}
# Goal is to determine firre abundance in wt & ko induced and control.
#TODO is this the short time course -- or all?

# Firre's expression in all 0 tinepoints
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  # here is the filter to time point 0 
  filter(cell_type == "ESC", timepoint == 0)

# taking the average and std deviation of all 0 timepoints
firre_mean <- firre_tpm %>%
  # grouping by induced and genotype (firre_ko - col label)
  group_by(firre_induced, firre_ko) %>%
  summarize(sd = sd(tpm),
            tpm = mean(tpm))

# setting Firre wild-type average expression (25.1 in WT control line - not transge (45tpm))
firre_wt_level <- firre_mean$tpm[firre_mean$firre_induced == "control" & firre_mean$firre_ko == "WT"]

# Plotting mean Firre in KO, WT and induced at 0 timepoint 
ggplot(firre_tpm, aes(x = firre_induced, y = tpm)) +
  geom_beeswarm(cex = 6, alpha = 0.8) +
    stat_summary(aes(y = tpm, group = firre_induced), fun=mean, colour="red", geom="point") +
    stat_summary(fun='mean', geom='text', 
               aes(label=signif(..y..,3),x=firre_induced)) +
  facet_grid(~firre_ko)

# ggsave("figures/firre_expression_at_zero.pdf", height = 2, width = 2)

```
### Result: Firre is slightly higher in the transgene WT line. (Tg could be leaky?) 
### Result: Firre is not expressed in KO lines or Tg (thus Tg is not leaky)



### ESC KO long timecourse
## ESC KO (rescue)

```{r Firre KO + transgene long time course (long), message=FALSE}

firre_ko_long_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  # This is the filter to only KO thus control line and inducible
  filter(cell_type == "ESC", firre_ko == "KO", timecourse_length == "long")

# Summarizing cntrol, induced lines and mean tpm in KO background
firre_ko_long_means <- firre_ko_long_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))

# setting Firre wild-type average expression (25.1 in WT control line - not transge (45tpm))
firre_wt_level <- firre_mean$tpm[firre_mean$firre_induced == "control" & firre_mean$firre_ko == "WT"]

# Setting up tpm relative to wt average. new col rel_wt
firre_means <- firre_means %>%
  mutate(rel_wt = tpm / firre_wt_level)

# Plotting the K0 long time course for control and induced lines
ggplot(firre_means, 
       aes(x = timepoint, y = rel_wt, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = rel_wt, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")

# ggsave("figures/firre_esc_ko_profile_long_relwt.pdf", height = 2, width = 2, useDingbats = FALSE)

```
### Result: Firre is induced in the KO background to ~17 fold wildtype levels 
### Result: Control line does not express Firre upon dox treament at anytime point


# Examples of genes induced at same time as Firre (Potential firre targets)
# This is a bit adhoc as we have not differential expression yet.
# Examples are in both WT and KO backgorunds with transgene induction.

# Rapgef4 expression levels Only for SHORT time course (0-330 min)
```{r 0-330 min : Rapgef4 in KO and WT + trans gene induction and dox control}

# filter to Rapgef4 tpms in wt and KO tiomecourse 0-330 min.
firre_tpm <- tpm %>%
  filter(gene_name == "Rapgef4") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")

# summarizing control and induced tpm means
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))

# plotting each all raw data and replicates for Rapgef4
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_wrap(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")

# ggsave("figures/rapgef4_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```
### Result: Rapgef4 is not expressed and activates txn in early time points 

# Adgrg1 expression levels Only for SHORT time course (0-330 min)
```{r 0-330 min : Adrgr1 in KO and WT + trans gene induction and dox control}

# Filtering to Adrgr1
firre_tpm <- tpm %>%
  filter(gene_name == "Adgrg1") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")

# summarizing control and induced tpm means
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))

# plotting each all raw data and replicates for Adgrg1
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")

# ggsave("figures/adgrg1_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```
### Result: Adgrg1 is not expressed and activates txn in early time points 


# Shf expression levels Only for SHORT time course (0-330 min)
```{r 0-330 min : Shf in KO and WT + trans gene induction and dox control}

# Filtering to Shf
firre_tpm <- tpm %>%
  filter(gene_name == "Shf") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")

# summarizing control and induced tpm means
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))

# plotting each all raw data and replicates for Shf
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")

# ggsave("figures/shf_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```
### Result: Shf is already expressed and goes up a lot quickly
### Result: Unlike Adgrg1 and Rapgef 4 this gene is already on and bootst up faster


# Gfap expression levels Only for SHORT time course (0-330 min)
```{r r 0-330 min : Gfap in KO and WT + trans gene induction and dox control}

# Filtering to Gfap
firre_tpm <- tpm %>%
  filter(gene_name == "Gfap") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")

# summarizing control and induced tpm means
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))

# plotting each all raw data and replicates for Gfap
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/gfap_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```
### Result: Gfap is also off and turns on Rapidly
#### Summary: Gfap, Rapgef4, Adgrg1 are all off and turn up to ~2tpm :: Shf is on and boost up fast



# Bringing in the genes differentially expressed genes across short time course
# This is the 30 final genes
```{r getting to the 30 genes temporally regulated across short time course}

# Filtering genes to one value (if sig in more than one time point)
load("/scratch/Shares/rinn/JR/Firre_time/analysis/01_setup/results/short_vszero_sig.RData", verbose = T)
short_sig_genes <- unique(short_vszero_sig$gene_id)

# Pulling in TPM values from TPM object and adding to short_sig_genes
short_sig_tpm <- tpm %>%
  filter(gene_id %in% short_sig_genes) %>%
    pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")

# Finding the max value of gene across time course.
short_sig_tpm_summary <- short_sig_tpm %>%
  group_by(gene_id, gene_name) %>%
  summarize(max_tpm = max(tpm))

# Now we have the max value for all 30 genes.
```

```{r}
firre_tpm <- tpm %>%
  filter(gene_name == "Dpf2") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/dpf2_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```

```{r}
firre_tpm <- tpm %>%
  filter(gene_name == "Nceh1") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced, firre_ko) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  facet_grid(~firre_ko) +
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/nceh1_short.pdf", height = 1.5, width = 3, useDingbats = FALSE)
```

### ESC KO short timecourse

```{r esc_ko_short_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "KO", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_ko_profile_short.pdf", height = 3, width = 3, useDingbats = FALSE)
# Label max
firre_means <- firre_means %>%
  mutate(rel_wt = tpm / firre_wt_level)
ggplot(firre_means, 
       aes(x = timepoint, y = rel_wt, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = rel_wt, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  geom_text(data = firre_means %>% filter(rel_wt == max(rel_wt)), aes(y = rel_wt + 2, label = round(rel_wt,1))) +
  scale_x_continuous(breaks = seq(0,330, by = 30), labels = c("0", "", "1", "", "2", "", "3", "", "4", "", "5", "")) +
  xlab("Time (h)")
ggsave("figures/firre_esc_ko_profile_short_relwt.pdf", height = 2, width = 2, useDingbats = FALSE)
```

### ESC KO Combined

```{r esc_ko_combined_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "KO")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_ko_profile.pdf", height = 3, width = 3, useDingbats = FALSE)
```

## ESC WT (overexpression)

### ESC WT long timecourse

```{r esc_wt_long_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "WT", timecourse_length == "long")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_wt_profile_long.pdf", height = 3, width = 3, useDingbats = FALSE)
firre_means <- firre_means %>%
  mutate(rel_wt = tpm / firre_wt_level)
ggplot(firre_means, 
       aes(x = timepoint, y = rel_wt, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = rel_wt, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  geom_text(data = firre_means %>% filter(rel_wt == max(rel_wt)), aes(y = rel_wt + 0.5, label = round(rel_wt,1))) +
  scale_x_continuous(breaks = c(0, 720, 1440, 2880, 5670), labels = c("0", "12", "24", "48", "96")) +
  xlab("Time (h)") +
  ylim(0,6)
ggsave("figures/firre_esc_wt_profile_long_relwt.pdf", height = 2, width = 2, useDingbats = FALSE)
```

### ESC WT short timecourse

```{r esc_wt_short_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "WT", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_wt_profile_short.pdf", height = 3, width = 3, useDingbats = FALSE)
firre_means <- firre_means %>%
  mutate(rel_wt = tpm / firre_wt_level)
ggplot(firre_means, 
       aes(x = timepoint, y = rel_wt, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = rel_wt, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  geom_text(data = firre_means %>% filter(rel_wt == max(rel_wt)), aes(y = rel_wt + 2, label = round(rel_wt,1))) +
  scale_x_continuous(breaks = seq(0,330, by = 30), labels = c("0", "", "1", "", "2", "", "3", "", "4", "", "5", "")) +
  xlab("Time (h)") +
  ylim(0,50)
ggsave("figures/firre_esc_wt_profile_short_relwt.pdf", height = 2, width = 2, useDingbats = FALSE)
```

### ESC WT Combined

```{r esc_wt_combined_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", firre_ko == "WT")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_esc_wt_profile.pdf", height = 3, width = 3, useDingbats = FALSE)
```

## NPC KO (rescue)

### NPC KO short timecourse

```{r npc_ko_short_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "NPC", firre_ko == "KO", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_npc_ko_profile_short.pdf", height = 3, width = 3, useDingbats = FALSE)
```

## NPC WT (overexpression)

### NPC WT short timecourse

```{r npc_wt_short_firre_profile, message=FALSE}
firre_tpm <- tpm %>%
  filter(gene_name == "Firre") %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "NPC", firre_ko == "WT", timecourse_length == "short")
firre_means <- firre_tpm %>%
  group_by(timepoint, firre_induced) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint)))
ggplot(firre_tpm, 
       aes(x = as.numeric(as.character(timepoint)), y = tpm, color = firre_induced)) +
  geom_line(data = firre_means, aes(x = timepoint, y = tpm, color = firre_induced,
                                    group = firre_induced)) +
  geom_point() + 
  theme(legend.position = "none") +
  xlab("T (min)")
ggsave("figures/firre_npc_wt_profile_short.pdf", height = 3, width = 3, useDingbats = FALSE)
```