---
title: "Short timecourse rescue"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../../util/_plot_theme.R")
source("../../util/_util.R")
source("../01_setup/assumptions.R")
source("../../util/figures.R")

# Keep track of thresholds for figure naming
# These values are comming from assumptions.R and can be changed there
thresh <- paste0("pval", pval_thresh, "_l2fc", round(l2fc_thresh, 2))
```

```{r load, include=FALSE}

#TODO how to compile an .RData

# This was run on Pipeline nf-core/rnaseq -r 1.4.2 \
# Key data values we compiled into .RData file 

# Loading in .RData file of rnaseq data that includes
# Counts, G2S (genes to symbol), genes
# Salmon gene-counts :: analysis done on gene level (not isoform)
# Samples with information on Fastq files
# TPM values from DeSeq and tx2gene

load("../01_setup/results/rnaseq_data.RData")


```

# Here we will analyze the Firre KO bacground containing a Firre Transgene (tg) from 30-330min)
# We will also compare a Firre KO line that CANNOT induce the Firre-tg  (KO cntrl)
# We will use the same linear models and LRT test between them as in 04_firre_kpo_long_timecourse


# Deseq on all short time course samples in KO background
# Result files will be combined for later filtering on sig DEG genes (Firre responders)
```{r Deseq KO shor_timecourse}
if(!file.exists("results/ko_rescue_short.RData")) {
  
  # Filter to ESC KO short timecourse in samples
  ko_rescue_short_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_ko == "KO")
  
  # Filtering out counts data to KO short time course
  ko_rescue_short_counts <- salmon_gene_counts[,ko_rescue_short_samples$sample_id]
  
  # Check ordering of samples and counts
  stopifnot(all(rownames(ko_rescue_short_samples) == colnames(ko_rescue_short_counts)))
  stopifnot(all(rownames(ko_rescue_short_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test between two linear models
  ko_rescue_short_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_short_counts, 
                                                colData = ko_rescue_short_samples, 
                                                design = ~ firre_induced + timepoint + timepoint*firre_induced)
  # Likelihood ratio test to control for Dox
  # Model where timepoint*firre_induced interaction term is more significant
  ko_rescue_short_dds <- DESeq(ko_rescue_short_dds, test = "LRT", reduced = ~ firre_induced + timepoint)
```

# Filtering out Deseq results (non shrunken lfc)

```{r  filter Deseq results (not shrunken lfc)}
# Filter and Compile results
  #TODO Better understand the result outputs
  # listing out all results (resultsNames)
  res_names <- resultsNames(ko_rescue_short_dds)
  
  # Taking the interaction term results (firre_inducedfirre_induced.timepoint)
  # NOTE: dynamic res will be indexed in all subsequent filtering (as names of each result = 11)
  dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]
  
  
  # filter result cols for each result name and merge at end
  ko_rescue_short_lfc <- lapply(dynamic_res, function(x) {
    results(ko_rescue_short_dds, 
            # taking in each result name as x convert to df
            name = x) %>%
      as.data.frame() %>%
      # meta gene_id -> this will is to merge each but after bind rows is not meta
      rownames_to_column("gene_id") %>% 
      # getting gene name (G2s) and merging in 
      merge(g2s) %>%
      # changing name of results to timepoint
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
    # merging the "for loop-ish" function results
  }) %>% bind_rows()
```

# Filtering Deseq result files and generating shrunken lfc with apeglm package/function
```{r filter Deseq results (not shrunken lfc)}
# Shrunken LFC results generated by apeglm function
#TODO why do this and not use Deseq?
  # NOTE takes awhile for apeglm to run on 11 smaples !
  
  ko_rescue_short_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(ko_rescue_short_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()

```

# Filtering to maximum fold change (the timepoint of max fold change)
```{r filter Deseq results (not shrunken lfc)}

# Calculate the maximum fold-change in any one timepoint
#TODO is this one value == max fc timepoint?

  ko_rescue_short_maxfc <- ko_rescue_short_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  ko_rescue_short_shrnklfc <- ko_rescue_short_shrnklfc %>%
    left_join(ko_rescue_short_maxfc)
  
  save(ko_rescue_short_lfc, ko_rescue_short_shrnklfc, file = "results/ko_rescue_short.RData")
}

load("results/ko_rescue_short.RData")
```

# Deseq on all short time points versus 0  timepoint 
# Fold changes vs zero timepoint not using LRT with control line
```{r running Deseq on KO short time course}

# testing to see if file exists and if not run Desq
if(!file.exists("results/ko_rescue_short_vs_zero.RData")) {
  
  ko_rescue_short_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_ko == "KO",
           firre_induced == "firre_induced")
  
  # sub-set to gene level salmon counts (could also be redone with isofroms)
  ko_rescue_short_vszero_counts <- salmon_gene_counts[,ko_rescue_short_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(ko_rescue_short_vszero_samples) == colnames(ko_rescue_short_vszero_counts)))
  stopifnot(all(rownames(ko_rescue_short_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  ko_rescue_short_vszero_dds <- DESeqDataSetFromMatrix(countData = ko_rescue_short_vszero_counts,
                                                       colData = ko_rescue_short_vszero_samples,
                                                       design = ~ timepoint)
  ko_rescue_short_vszero_dds <- DESeq(ko_rescue_short_vszero_dds)
  res_names <- resultsNames(ko_rescue_short_vszero_dds)
  
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  ko_rescue_short_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(ko_rescue_short_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  save(ko_rescue_short_vszero_shrnklfc, file = "results/ko_rescue_short_vs_zero.RData")
}

load("results/ko_rescue_short_vs_zero.RData")
```

### Rescued genes

These are the genes that are reciprocally regulated in the KO vs WT and the Firre induced line.

```{r}
# We need the KO vs WT data
load("../04_firre_ko_vs_wt/results/wtko.RData")

wtko_res_shrnklfc <- wtko_res_shrnklfc %>%
  mutate(sig = padj <= pval_thresh & abs(log2FoldChange) > l2fc_thresh)

wtko_sig <- wtko_res_shrnklfc %>% 
  filter(sig == TRUE)

# Let's figure out which genes are rescued
rescued_genes <- wtko_sig$gene_id[wtko_sig$gene_id %in% 
                                    (ko_rescue_short_shrnklfc %>%
                                       filter(padj <= pval_thresh) %>% 
                                       pull(gene_id))]

rescue_wtko <- wtko_sig %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange) %>%
  dplyr::rename(l2fc_wtko = log2FoldChange)

rescue_ltc <- ko_rescue_short_vszero_shrnklfc %>%
  filter(gene_id %in% rescued_genes) %>%
  dplyr::select(gene_id, gene_name, log2FoldChange, timepoint) %>%
  dplyr::rename(l2fc_ltc = log2FoldChange)

rescued_df <- rescue_wtko %>%
  left_join(rescue_ltc)

rescued_df <- rescued_df %>%
  mutate(dist = abs((l2fc_wtko * -1) - l2fc_ltc))

rescued_df_closest <- rescued_df %>%
  group_by(gene_id, gene_name) %>%
  mutate(closest = min(dist) == dist,
         max_fc = max(abs(l2fc_ltc)) == abs(l2fc_ltc)) %>%
  filter(closest == TRUE)

rescued_df_closest <-  rescued_df_closest %>%
  mutate(twenty_percent = l2fc_wtko * 0.2,
         rescued = (-1*l2fc_wtko - twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko + twenty_percent) |
           (-1*l2fc_wtko + twenty_percent)>= l2fc_ltc & l2fc_ltc>= (-1*l2fc_wtko - twenty_percent))

rescued_genes_20 <- rescued_df_closest %>% 
  filter(rescued == TRUE) %>% 
  pull(gene_name)

ggplot(rescued_df_closest, aes(x = l2fc_wtko, y = l2fc_ltc, color = rescued)) +
  geom_abline(slope = -1) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point(alpha = 0.7) +
  geom_abline(slope = -0.8, lty = 2) +
  geom_abline(slope = -1.2, lty = 2) +
  xlim(-2,2) +
  ylim(-2,2)
ggsave(paste0("figures/short_rescue_lfc_scatter_", thresh, ".pdf"), 
       height = 3, width = 3, useDingbats = FALSE)
```

### Short timecourse call significant genes

We'll make the p-value cutoff based on the dox controlled model and the l2fc cutoff
based on the fold change vs zero.

```{r}
ko_rescue_short_dox_sig <- ko_rescue_short_shrnklfc %>% 
  filter(padj <= pval_thresh)

ko_rescue_short_vszero_sig <- ko_rescue_short_vszero_shrnklfc %>%
  filter(gene_id %in% ko_rescue_short_dox_sig$gene_id)

ko_rescue_short_vszero_maxfc <- ko_rescue_short_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

ko_rescue_short_vszero_sig <- ko_rescue_short_vszero_sig %>%
  left_join(ko_rescue_short_vszero_maxfc)

ko_rescue_short_vszero_sig <- ko_rescue_short_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

save(ko_rescue_short_vszero_sig, file = "results/ko_rescue_short_vszero_sig.RData")
```

```{r}
load("results/ko_rescue_short_vszero_sig.RData", verbose = T)
ko_rescue_short_vszero_sig <- ko_rescue_short_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)
make_timecourse_lfc_plot(deseq_res = ko_rescue_short_vszero_sig,
                         genes_to_include = unique(ko_rescue_short_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/ko_rescue_short_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)

deseq_res <- ko_rescue_short_vszero_sig
genes_to_include <- unique(ko_rescue_short_vszero_sig$gene_id)
# Let's plot the downregulated genes separately
 deg <- deseq_res %>%
    filter(gene_id %in% genes_to_include) %>%
    dplyr::select(gene_id, gene_name, baseMean, l2fc_shrunken, timepoint)
  
  # Add a zero timepoint
  deg_zero <- deg %>%
    dplyr::select(gene_id, gene_name, baseMean) %>%
    distinct() %>%
    mutate(l2fc_shrunken = 0,
           timepoint = 0) 
  
  deg <- deg %>%
    bind_rows(deg_zero)
  
  if(length(unique(deg$gene_id)) > 1) {
    deg <- deg %>%
      mutate(x_adjust = scale(log10(baseMean)),
             tp_adjust = timepoint + x_adjust*x_adjust_scale)
  } else {
    deg <- deg %>%
      mutate(tp_adjust = timepoint)
  }
  
  num_deg <- deg  %>%
    group_by(gene_id) %>%
    summarize(max_fc = l2fc_shrunken[which.max(abs(l2fc_shrunken))]) %>%
    mutate(direction = ifelse(max_fc > 0, "up", "down")) %>%
    group_by(direction) %>%
    summarize(num_deg = length(unique(gene_id)))
  
  label_x <- max(deg$timepoint / 2)
  
  # Let's sort on the order of fold-change
  deg_fc <- deg %>%
    group_by(gene_name) %>%
    summarize(max_fc = l2fc_shrunken[which.max(abs(l2fc_shrunken))]) %>%
    arrange(-max_fc)
  
  deg$gene_name <- factor(deg$gene_name, levels = deg_fc$gene_name)
  
  g1 <- ggplot(deg, aes(x = timepoint, y = l2fc_shrunken, group = gene_id)) +
    geom_hline(yintercept = 0) +
    geom_line(alpha = 0.2, color = "gray70") +
    geom_point(alpha = 0.4) +
    facet_wrap(~gene_name, scales = "free_y") +
    scale_x_continuous(breaks = seq(from = 0, to = 330, by = 30), labels = c("0", "", "1", "", "2", "", "3", "", "4", "", "5", "" ) )
  show(g1) 
ggsave("figures/short_rescue_individual_fc_profiles.pdf", device = cairo_pdf,
       height = 3.5, width = 4.5)
```

```{r}
# How many genes are changing in each timepoint?
deg_per_timepoint <- ko_rescue_short_vszero_sig %>%
  filter(abs(l2fc_shrunken) > l2fc_thresh) %>%
  mutate(direction = ifelse(l2fc_shrunken > 0, "up", "down")) %>%
  group_by(timepoint, direction) %>%
  summarize(ndeg = n()) %>%
  mutate(y = ifelse(direction == "up", ndeg, ndeg*-1))

ggplot(deg_per_timepoint, aes(x = timepoint, y = y, fill = direction)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "up"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=-0.25) +
  geom_text(data = deg_per_timepoint %>% filter(direction == "down"), aes(label=ndeg), position=position_dodge(width=0.9), vjust=1.1) +
    scale_x_continuous(breaks = seq(from = 0, to = 330, by = 30), labels = c("0", "", "1", "", "2", "", "3", "", "4", "", "5", "" ) ) +
  guides(fill = FALSE) +
  ylim(-6,15)
ggsave("figures/short_rescue_ndeg_per_timepoint.pdf", device = cairo_pdf,
       height = 2, width = 2.5)  
```



### Short timecourse Firre responders heatmap

```{r}
# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(ko_rescue_short_vszero_sig$gene_id)) == length(unique(ko_rescue_short_vszero_sig$gene_name))))

ko_rescue_short_lfc <- ko_rescue_short_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()

# Add a zero column.
ko_rescue_short_lfc <- cbind(matrix(0, nrow = nrow(ko_rescue_short_lfc), ncol = 1), ko_rescue_short_lfc)
colnames(ko_rescue_short_lfc)[[1]] <- "0"

row_ha = rowAnnotation(sig_in_ko = as.numeric(rownames(ko_rescue_short_lfc) %in% wtko_sig$gene_name),
                       rescued = as.numeric(rownames(ko_rescue_short_lfc) %in% rescued_genes_20),
                       col = list(sig_in_ko = c("1" = "black", "0" = "white"),
                                  rescued = c("1" = "black", "0" = "white")))

table(as.numeric(rownames(ko_rescue_short_lfc) %in% wtko_sig$gene_name))
table(as.numeric(rownames(ko_rescue_short_lfc) %in% rescued_genes_20))
rownames(ko_rescue_short_lfc)[rownames(ko_rescue_short_lfc) %in% rescued_genes_20]
pdf(paste0("figures/short_firre_responders_heatmap_", thresh, ".pdf"), 
    width = 4, height = 3.5)
h1 <- Heatmap(ko_rescue_short_lfc, 
        name = "l2fc",
        cluster_columns = FALSE, show_row_names = FALSE, 
        col = colorRamp2(seq(-2,2,length.out = 100), col_pal10),
        left_annotation = row_ha)
draw(h1)
dev.off()
draw(h1)
```