---
title: "H3K27me3 ChIP-seq"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Rsubread)
library(DESeq2)
source("../../util/_plot_theme.R")
# source ~/anaconda3/etc/profile.d/conda.sh
library(ChIPseqSpikeInFree)


# #TODO what are these
# Sys.setenv(PATH = "/opt/gcc/7.2.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/usr/lib/rstudio-server/bin/postback")
# Sys.setenv(LD_LIBRARY_PATH = "/opt/gcc/7.2.0/bin:/usr/lib64/R/lib::/lib:/usr/lib/jvm/jre/lib/amd64/server:/usr/lib/jvm/jre/lib/amd64:/usr/lib/jvm/java/lib/amd64:/usr/java/packages/lib/amd64:/lib:/usr/lib")
# Sys.setenv(LD_LIBRARY_PATH = "/usr/local/lib:/usr/lib:/usr/local/lib64:/usr/lib64")
# Sys.getenv("PATH")
# Sys.getenv("LD_LIBRARY_PATH")
# #TODO what is this doesn't install well
# install.packages("xgboost")
# system("gcc --version")
# system('export CXXFLAGS="$CXXFLAGS -fPIC"')
# system('export R_LIBS="~/R/lib"')

```
# We performed K27me3 Chipseq at 48 hours after Firre induction in WT and KO lines and controls
# We wanted to see if K27me3 was lost or gained globally upon induction of Firre
# some evidence suggests that lack of Firre would result in decreased K27me3, thus induction would increase k37me3


# Running macs 
```{bash}
# needs to be installed in home directory
#TODO install anaconda
# Install Miniconda3 to run a python environment where macs3 is installed

# wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
# chmod u+x Miniconda3-latest-Linux-x86_64.sh 
# Users/jori2700/bin/Minconda3
# modified      /Users/jori2700/.bashrc
# add to path : .bash_profile 
# source .bash_profile
# conda activate base
# to deactivate the env session: conda deactivate

# create environment for macs3 with naming first
# conda create --name macs3

# need to update python to 3.8
# removing previous env : conda env remove --name macs3
# conda create --name macs3 python=3.8
#  source activate macs3
# pip install macs3


# Note Miniconda is now my new python environment so if other packages
# reequire python (eg. deeptools) create environment in there
```

# Setting up SLURM job to run macs3
```{bash running macs3 with SLURM}

# Combine all data and call peaks on everything combined.
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=h3k27me3_chipseq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=john.rinn@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=10:00:00
#SBATCH --output=h3k27me3_chip_macs.out
#SBATCH --error=h3k27me3_chip_macs.err
date; hostname

# going to nf_core chipseq outputs
BASE_PATH=/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary

# calling out macs3 to avtivate an env
source activate macs3

# callpeak to call peaks
# -f means bam file paired end 
# -t is treatment group / -c is control group
# -g is the genome "mm"
# -n is the name of the output file
# -- broad peak (narrow peak is the default)
# --broad-cutoff is distance between peaks to be merged
# -B probably makes a bam file for the outputs

macs3 callpeak -f BAMPE -t \
${BASE_PATH}/KO_CTL_48_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KO_CTL_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KOiTg_48_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KOiTg_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WT_CTL_48_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WT_CTL_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WTiTg_48_IP_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WTiTg_IP_R1.mLb.clN.sorted.bam \
-c \
${BASE_PATH}/KO_CTL_48_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KO_CTL_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KOiTg_48_R1.mLb.clN.sorted.bam \
${BASE_PATH}/KOiTg_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WT_CTL_48_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WT_CTL_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WTiTg_48_R1.mLb.clN.sorted.bam \
${BASE_PATH}/WTiTg_R1.mLb.clN.sorted.bam \
-g mm -n FIRRE_H3K27me3_chipseq_combined --broad --broad-cutoff 0.1 -B
date


```


# Importing ChIPseq peak files from MACS run NF_CORE chipseq pipeline
# Running feature counts over peaks for Deseq input values

```{r importing chip seq files and quantitating with Feature counts}

# reading in macs3 outpout broad.peak file
merged_peaks <- rtracklayer::import("macs/FIRRE_H3K27me3_chipseq_combined_peaks.broadPeak")

# making into a dataframe
merged_peaks_df <- merged_peaks %>% as.data.frame()

# this is the input for feature counts: creating .saf file
merged_peaks_saf <- merged_peaks_df %>%
  dplyr::select(name,seqnames, start, end, strand) %>%
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand)

write.table(merged_peaks_saf, "data/merged_peaks.saf", quote = F, row.names = F, sep = "\t")

#TODO more stable file path
# collecting all the bams for K27me3 chipseq to be used in feature counts
bam_files <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/", pattern = ".bam$", full.names = T)

# using feature counts to quantitate K27me3 peaks in Firre induced and control lines

# Run this the first time - in the future load in .rds

### IMPORTANT ###
# counts <- featureCounts(bam_files, 
#                         annot.ext = "data/merged_peaks.saf", 
#                         isPairedEnd = TRUE, nthreads = 16)

# Saving counts object for future analyses
# write_rds(counts, "results/merged_peak_counts.rds")

# test read in of .rds
counts <- read_rds("results/merged_peak_counts.rds")

# checking length of peaks with counts
nrow(counts$counts)
# 102,957 peaks in merged data set
```

# Creating smaples sheet matrix for Deseq
```{r Creating experimental matrix for Deseq of Firre induced vs cntrl K27me3 ChipSeq}

# laoding in counts from feature counts (above)

# test read in of .rds
counts <- read_rds("results/merged_peak_counts.rds")

# reading in counts for each peaks as matrix
count_matrix <- counts$counts
# counts matrix has 102,957 peaks (rows) and 16 cols for samples

# reading in sample sheet
samples <- read.csv("data/h3k27me3_firre_samplesheet.csv")

# changing row names to match
rownames(samples) <- samples$sample_id

# checking everything is in order
stopifnot(all(colnames(count_matrix) == samples$bam_file))

# TODO the baove doesn't match so added col names to make it match 
# changing col names to match with samples 
colnames(count_matrix) <- samples$sample_id

# checking everything is in order
stopifnot(all(colnames(count_matrix) == samples$sample_id))

# merging fire_itg and timepoint values into new col called itg_timepoint
samples <- samples %>% unite(itg_timepoint, firre_itg, timepoint, remove = FALSE)

# making a new control col
samples$condition <- "control"

# setting control col values to induced if == ITG_28 
samples[samples$itg_timepoint == "ITG_48", "condition"] <- "induced"

# grabbing IP samples for 48hrs and then indexing the counts with those samples
ip_samples <- samples %>% filter(h3k27me3 == "IP", timepoint == 48)
ip_counts <- count_matrix[,ip_samples$sample_id]
```

# Running Deseq on comparisons of induced firre versus control for each time point
# Determining how many peaks are significant between induced and control lines
```{r Running Deseq on condition of peaks based on Firre induction}

# setting up experiment matrix
# with out factoring firre_itg it's alphabetical
# let's factor induced vs control 

ip_samples$firre_itg <- factor(ip_samples$firre_itg, levels = c("CTL", "ITG"))

# setting up deseq
dds <- DESeqDataSetFromMatrix(countData = ip_counts,
                              colData = ip_samples,
                              design = ~ firre_itg)
# run deseq
dds <- DESeq(dds)

# checking results
resultsNames(dds)
# removing intercept results by using "results" function to "grep" firre_itg ...
res <- results(dds, name = "firre_itg_ITG_vs_CTL")

# turning into data frame
res_df <- res %>% as.data.frame()

# filtering to only peaks that are sig differnet of P < 0.05
sig <- res_df %>% filter(padj < 0.05)

# plotting mean vs fold change (MA plot is typically log10)
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# Plotting pval vs fold change
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result: no significant peaks were found (P ~ .17 for the most sig)
### Result: Firre induction did not increase or decrase K27me3 at 48hrs


# Now just wild-type IP sampels for induced versus control at 28hrs
```{r comparing WT control versus firre induced for sig peaks via Deseq}

# selecting just wt_ip samples and then indexing to get counts
wt_ip_samples <- samples %>% filter(firre_ko == "WT", h3k27me3 == "IP")
wt_ip_counts <- count_matrix[,wt_ip_samples$sample_id]

# setting factor levels
wt_ip_samples$condition <- factor(wt_ip_samples$condition, levels = c("control", "induced"))

# Setting up Deseq design on 
dds <- DESeqDataSetFromMatrix(countData = wt_ip_counts,
                              colData = wt_ip_samples,
                            design = ~ condition)

# running Deseq
dds <- DESeq(dds)

# checking results and filtering
resultsNames(dds)
res <- results(dds, name = "condition_induced_vs_control")

# creating a data frame of sig peaks
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Ploitting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result no significant peaks found in WT induced versus 


# Testing KO cntrol vs induced by Deseq for sig peaks
```{r Sig peaks between KO cntrl and induced}

# Filtering to just KO samples and then indexing to get counts for peaks
ko_ip_samples <- samples %>% filter(firre_ko == "KO", h3k27me3 == "IP")
ko_ip_counts <- count_matrix[,ko_ip_samples$sample_id]

# factoring
ko_ip_samples$condition <- factor(ko_ip_samples$condition, levels = c("control", "induced"))

# setting up Deseq
dds <- DESeqDataSetFromMatrix(countData = ko_ip_counts,
                              colData = ko_ip_samples,
                             design = ~ condition)

# Running Deseq
dds <- DESeq(dds)

# getting names and filtering to sig
resultsNames(dds)
res <- results(dds, name = "condition_induced_vs_control")

# sig peaks dataframe
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result no sig peaks

# Sig peaks between 0 and 48 hours
```{r Deseq on timepoint of 0 vs 48 hours with both KO and WT mixed}

# 48 itg vs 0 itg filtering and getting counts
itg_ip_samples <- samples %>% filter(h3k27me3 == "IP", firre_itg == "ITG")
itg_ip_counts <- count_matrix[,itg_ip_samples$sample_id]

# setting factors
itg_ip_samples$timepoint <- factor(itg_ip_samples$timepoint, c("0", "48"))



# setting up Deseq
dds <- DESeqDataSetFromMatrix(countData = itg_ip_counts,
                              colData = itg_ip_samples,
                              design = ~ timepoint)

# running Deseq
dds <- DESeq(dds)

# getting results and filtering to sig diff peaks
resultsNames(dds)
res <- results(dds, name = "timepoint_48_vs_0")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result: no sig peaks between time points


# Comparing control lines for sig peaks. 
```{r comparing control lines WT and KO between 0 and 48 hours}
# 48 ctl vs 0 ctl

# filtering to samples and indexing to get counts for peaks
ctl_ip_samples <- samples %>% filter(h3k27me3 == "IP", firre_itg == "CTL")
ctl_ip_counts <- count_matrix[,ctl_ip_samples$sample_id]

# setting factors
ctl_ip_samples$timepoint <- factor(ctl_ip_samples$timepoint, c("0", "48"))

# setting up Deseq
dds <- DESeqDataSetFromMatrix(countData = ctl_ip_counts,
                              colData = ctl_ip_samples,
                             design = ~ timepoint)

# Running
dds <- DESeq(dds)

# filtering results and to sig peaks
resultsNames(dds)
res <- results(dds, name = "timepoint_48_vs_0")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()

```
### Result: No sig peaks

# *** ALL WT vs KO at 0 timepoint : HAS PEAKS worth looking into! ***
# Taking only the control samples and IP = WT vs KO for all control samples and thus just WT vs KO 

```{r Control samples only WT -vs- KO sig peaks}

# filtering samples and indexing to get counts
zero_samples <- samples %>% filter(h3k27me3 == "IP", condition == "control")
zero_counts <- count_matrix[,zero_samples$sample_id]

# setting factors to compare to wt
zero_samples$firre_ko <- factor(zero_samples$firre_ko, c("WT", "KO"))

# setting up Deseq
dds <- DESeqDataSetFromMatrix(countData = zero_counts,
                              colData = zero_samples,
                             design = ~ firre_ko)

# running 
dds <- DESeq(dds)

# filtering results and getting sig peaks
resultsNames(dds)


res <- results(dds, name = "firre_ko_KO_vs_WT")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
# HUH this shows that KO seems to have more k27me3



# finding sig peaks with LFC > 2 fold up or down

# loss of K27me3 in KO 
### NOTE not sigificant they are fold change is happening 
wt_vs_ko_k27me3_sig_down <- res_df %>% filter(log2FoldChange < -1)
write_csv(wt_vs_ko_k27me3_sig_down, "results/wt_vs_ko_k27me3_sig_down.csv")

# gain of K27me3 in KO 
wt_vs_ko_k27me3_sig_up <- res_df %>% filter(log2FoldChange > 1)
write_csv(wt_vs_ko_k27me3_sig_up, "results/wt_vs_ko_k27me3_sig_up.csv")

# combined sig
all_wt_vs_ko_sig_k27me3 <- bind_rows(wt_vs_ko_k27me3_sig_up, wt_vs_ko_k27me3_sig_down)
write_csv(all_wt_vs_ko_sig_k27me3, "results/all_wt_vs_ko_sig_k27me3.csv")


# filter versus indexing == get to know it
# filter likes a data frame and then can index into cols for logical tests

# THIS IS THE WAY
#TODO why is there a logical error with these types of indexing
sig_threshold <- sig[(sig$log2FoldChange >= 1),]
# NO sig peaks at this threshold 


sig_threshold <- filter(sig[sig$log2FoldChange] >= -1)
sig_x <- res_df %>%
  filter(abs(log2FoldChange > 1))
sig_threshold <- filter(sig$log2FoldChange <= -1)

#TODO figure out how to connect peak names to big wig and look at raw dat for these sites
#TODO do the sig peaks overlap target genes.
```
### Result: there are some significant peaks between WT and KO worth looking at more !
### Result: the fold changes are not that big -- look at "sig" peaks in raw data
### Result: there are only 115 sig peaks so not very maany relative to  ~120,000 peaks in counts





# Deseq on the 48 hr timepoint comparing all 48 hr time points
# Above tested for the 0 timepoint essentially WT vs KO
# Now we are doing WT vs KO after exposure to firre combined with controls (e.g. WT)  same vs KO
# Note that this is confounding induced and control models - the above is more realistic WT vs KO
```{r 48 timepoint only design model KO vs WT}

# grabbing 48 hr samples
fortyeight_samples <- samples %>% filter(h3k27me3 == "IP", timepoint == 48)
fortyeight_counts <- count_matrix[,fortyeight_samples$sample_id]

# setting up factors
fortyeight_samples$firre_ko <- factor(fortyeight_samples$firre_ko, c("WT", "KO"))

# setting up Deseq
dds <- DESeqDataSetFromMatrix(countData = fortyeight_counts,
                              colData = fortyeight_samples,
                              design = ~ firre_ko)
# running 
dds <- DESeq(dds)
resultsNames(dds)

# filtering to sig P < 0.05 peak differences
res <- results(dds, name = "firre_ko_KO_vs_WT")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# there are 116 sig peaks down 2 fold
sig_up <- res_df %>% filter(log2FoldChange < -1)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()

#TODO check out this plot
#TODO combined LRT linear model test
```
### Result: There is a global trend of less K27m3 in Firre KO lines
### Result: there are 116 sig down peaks > 2 fold -- but kinda  confusing  set up WT vs KO  0 time point is prob better


# Comparing just the control lines across all timepoints
```{r control lines across all timepoints}

# selecting control lines and setting levels to WT and KO 
ip_samples <- samples %>% filter(h3k27me3 == "IP", condition == "control")
ip_counts <- count_matrix[,ip_samples$sample_id]

ip_samples$firre_ko <- factor(ip_samples$firre_ko, levels = c("WT", "KO"))

# Running Deseq
dds <- DESeqDataSetFromMatrix(countData = ip_counts,
                              colData = ip_samples,
                              design = ~ firre_ko)

# getting results and filtering to sig P < 0.05
dds <- DESeq(dds)
resultsNames(dds)
res <- results(dds, name = "firre_ko_KO_vs_WT")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()


# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result: The first time we see increased K27me3 (check design WT/KO?)
### AGain seeing KO has more k27me not sig 


# Trying a combined model to determine peaks that are not in any dox control but in 
```{r LRT for dox effects and determing }

# selecting samples
ip_samples <- samples %>% filter(h3k27me3 == "IP")
ip_counts <- count_matrix[,ip_samples$sample_id]

# setting factors
#TODO what is going on here ?
ip_samples$firre_ko <- factor(ip_samples$firre_ko, levels = c("WT", "KO"))
ip_samples$timepoint <- factor(ip_samples$timepoint, levels = c("0", "48"))
dds <- DESeqDataSetFromMatrix(countData = ip_counts,
                              colData = ip_samples,
                              design = ~ firre_itg + timepoint + firre_itg*timepoint)

# running Deseq
dds <- DESeq(dds, test = "LRT",
             reduced = ~ firre_itg + timepoint)

# getting results and filtering to P < 0.05
resultsNames(dds)
res <- results(dds, name = "firre_itgITG.timepoint48")
res_df <- res %>% as.data.frame()
sig <- res_df %>% filter(padj < 0.05)

# Plotting base mean vs fold change
ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + geom_point()

# plotting fold change versus pval
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```
### Result: seeing increased K27me3 in WT controlled for dox? (DOBLE CHECK MODEL)


# All peak files
```{r}

#TODO start here
# collect the peakfiles
peak_files <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks", pattern = ".broadPeak", full.names = T)
cat(peak_files)
```


# Performing ChIPr for more stringen peak differences.
```{bash}
# CHIPR
# source ~/anaconda3/bin/activate chipr

chipr -i /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/KO_CTL_48_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/KO_CTL_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/KOiTg_48_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/KOiTg_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/WT_CTL_48_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/WT_CTL_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/WTiTg_48_IP_R1_peaks.broadPeak /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/broad_peaks/WTiTg_IP_R1_peaks.broadPeak -m 2 -o h3k27me3_ip_chipr
```

```{r}
# Peaks passing IDR
chipr_peaks <- read.table("data/chipr/h3k27me3_ip_chipr_optimal.bed") 
names(chipr_peaks) <- c("chr",	"start", "end", "name", "score", "strand", "signalValue", "pval", "qval")

chipr_bed <- chipr_peaks %>% dplyr::select(chr, start, end) %>% filter(grepl("chr", chr))
write.table(chipr_bed, "results/chipr_peaks.bed", sep = "\t", quote = F, col.names = F, row.names = F)

sig_chipr_peaks <- chipr_peaks %>% filter(qval < 0.05)

chipr_peaks_gr <- GRanges(seqnames = chipr_peaks$chr,
                          ranges = IRanges(start = chipr_peaks$start,
                                           end = chipr_peaks$end))


chipr_merge_ov <- findOverlaps(chipr_peaks_gr, merged_peaks)

table(1:length(chipr_peaks_gr) %in% unique(chipr_merge_ov@from))



```


# TRACK HUB ?? Make seperate .RMD?
# Get the data ready for UCSC trackhub

chrom - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).
chromStart - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered 0.
chromEnd - The ending position of the feature in the chromosome or scaffold. The chromEnd base is not included in the display of the feature, however, the number in position format will be represented. For example, the first 100 bases of chromosome 1 are defined as chrom=1, chromStart=0, chromEnd=100, and span the bases numbered 0-99 in our software (not 0-100), but will represent the position notation chr1:1-100. Read more here.
The 9 additional optional BED fields are:

name - Defines the name of the BED line. This label is displayed to the left of the BED line in the Genome Browser window when the track is open to full display mode or directly to the left of the item in pack mode.
score - A score between 0 and 1000. If the track line useScore attribute is set to 1 for this annotation data set, the score value will determine the level of gray in which this feature is displayed (higher numbers = darker gray). This table shows the Genome Browser's translation of BED score values into shades of gray:
shade	 	 	 	 	 	 	 	 	 
score in range  	≤ 166	167-277	278-388	389-499	500-611	612-722	723-833	834-944	≥ 945
strand - Defines the strand. Either "." (=no strand) or "+" or "-".

```{r}
# http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64.v385/bedToBigBed
# http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes
fl <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/macs/broadPeak", 
                 full.names = T, pattern = ".broadPeak")
bp <- lapply(fl, rtracklayer::import)
names(bp) <- gsub("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/macs/broadPeak/", "", fl)
# Filter to canonical chrs
chrs <- c(paste0("chr", 1:19), "chrM", "chrX", "chrY")
for(i in 1:length(bp)) {
  bp[[i]] <- bp[[i]][seqnames(bp[[i]]) %in% chrs]
}

bpdf <- lapply(bp, as.data.frame)
# Export

for(i in 1:length(bp)) {
  write.table(bpdf[[i]] %>% dplyr::select(seqnames, start, end, name, score, strand, signalValue, pValue, qValue) %>%
                mutate(strand = "."),
              paste0("results/individual_peak_calls/broadPeak/", names(bp)[i]),
              quote = F, row.names = F, col.names = F)
}

f <- data.frame(files = list.files("results/individual_peak_calls/broadPeak", full.names = T, pattern = "broadPeak")) %>%
  mutate(rep_name = gsub("results/individual_peak_calls/broadPeak/|_peaks.broadPeak", "", files),
         command = paste0(">&2 echo '", gsub("-", "_", rep_name), "\n'; bedToBigBed -as=fmt.as -type=bed6+3 -extraIndex=signalValue,pValue,qValue ../results/individual_peak_calls/broadPeak/", rep_name, "_peaks.broadPeak mm10.chrom.sizes ../results/individual_peak_calls/bigBed/", rep_name, ".bb")) %>%
  dplyr::select(command) %>%
  write.table("bin/broadpeak2bb.sh", quote = F, col.names = F, row.names = F)

```

```{bash}
chmod u+x broadpeak2bb.sh
./broadpeak2bb.sh

rsync results/individual_peak_calls/bigBed/* mism6893@biof-trackhub.colorado.edu:/srv/http/biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_prc2_chipseq/.
```


#TODO Is this track hub suff
```{r}

bb_fl <- list.files("results/individual_peak_calls/bigBed")
conditions <- gsub("_IP_R1.bb", "", bb_fl)
bw_fl <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/bigwig", pattern = ".bigWig")


i <- 2
track_lines <- c()
for(i in 1:length(conditions)){
  
  
  tfl_bb <- bb_fl[grep(paste0(conditions[i], "_IP"),bb_fl)]
  
  ip_track <- bw_fl[grep(paste0(conditions[i], "_IP"), bw_fl)]
  input_track <- bw_fl[grepl(paste0(conditions[i], "_R1"), bw_fl)]
  
  
  names(tfl_bb) <- gsub("_R1.bb", "", tfl_bb)
  names(input_track) <- gsub("_R1.bigWig", "", input_track)
  names(ip_track) <- gsub("_R1.bigWig", "", ip_track)
  
  
  
  
  
  # Make the IP track
  ip <- paste0("track ", names(ip_track) ,"
bigDataUrl https://biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_h3k27me3_chipseq/", ip_track,"
shortLabel ", names(ip_track) ,"
longLabel ", names(ip_track) ,"
color 168,64,76
type bigWig
visibility full

")
  
  input <- paste0("track ", names(input_track) ,"_input
bigDataUrl https://biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_h3k27me3_chipseq/", input_track,"
shortLabel ", names(input_track) ,"
longLabel ", names(input_track) ,"
color 168,64,76
type bigWig
visibility full

")
  
  # Peaks track
  peaks <- paste0("track ", names(tfl_bb) ,"_peaks
bigDataUrl https://biof-trackhub.colorado.edu/rinnlab/micahel/firre_timecourse/firre_h3k27me3_chipseq/", tfl_bb,"
shortLabel ", names(tfl_bb) ,"_peaks
longLabel ", names(tfl_bb) ,"_peaks
color 168,64,76
type bigBed
visibility full

")
  track_lines <- c(track_lines, ip, input, peaks)
}

write_lines(track_lines, "results/th.txt")

```


















```{r}
# WTiTG_0
# WT_CTL_0
# WT_CTL_48

# /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/WTiTg_48_IP_R1.mLb.clN.sorted.bam
```




```{bash}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=h3k27me3_chipseq
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=30gb
#SBATCH --time=1:00:00
#SBATCH --output=h3k27me3_chip_macs.out
#SBATCH --error=h3k27me3_chip_macs.err

date; hostname
source ~/anaconda3/bin/activate macs3
macs3 callpeak -f BAMPE -t /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WTiTg_48_IP_R1.mLb.clN.sorted.bam -c /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WTiTg_48_R1.mLb.clN.sorted.bam -g mm -n wt_itg_h3k27me3_chipseq --broad --broad-cutoff 0.1 -B
date
```

```{bash}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=h3k27me3_chipseq_ko
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=30gb
#SBATCH --time=1:00:00
#SBATCH --output=h3k27me3_chip_macs.out
#SBATCH --error=h3k27me3_chip_macs.err

date; hostname
source ~/anaconda3/bin/activate macs3
macs3 callpeak -f BAMPE -t /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KOiTg_48_IP_R1.mLb.clN.sorted.bam -c /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KOiTg_48_R1.mLb.clN.sorted.bam -g mm -n ko_itg_h3k27me3_chipseq --broad --broad-cutoff 0.1 -B
date
```


```{bash}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=h3k27me3_chipseq_ctl
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=3:00:00
#SBATCH --output=h3k27me3_chip_macs_itg.out
#SBATCH --error=h3k27me3_chip_macs_itg.err

date; hostname
source ~/anaconda3/bin/activate macs3
macs3 callpeak -f BAMPE -t /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WT_CTL_48_IP_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WT_CTL_IP_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WTiTg_IP_R1.mLb.clN.sorted.bam -c /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WT_CTL_48_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WT_CTL_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/WTiTg_R1.mLb.clN.sorted.bam -g mm -n wt_ctl_h3k27me3_chipseq --broad --broad-cutoff 0.1 -B
date
```

```{bash}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=h3k27me3_chipseq_ctl_ko
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=10:00:00
#SBATCH --output=h3k27me3_chip_macs_itg.out
#SBATCH --error=h3k27me3_chip_macs_itg.err

date; hostname
source ~/anaconda3/bin/activate macs3
macs3 callpeak -f BAMPE -t /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KO_CTL_48_IP_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KO_CTL_IP_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KOiTg_IP_R1.mLb.clN.sorted.bam -c /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KO_CTL_48_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KO_CTL_R1.mLb.clN.sorted.bam /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results_all_individual_runs/bwa/mergedLibrary/KOiTg_R1.mLb.clN.sorted.bam -g mm -n ko_ctl_h3k27me3_chipseq --broad --broad-cutoff 0.1 -B
date
```





```{bash }
bw_files <- list.files("/scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/bigwig/", pattern = ".bigWig")
bw_files
names(bw_files) <- gsub("_R1.bigWig", "", bw_files)

for(i in 1:length(bw_files)) {
  script <- paste0("#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=deeptools_profile", i, "
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --mem=20gb
#SBATCH --time=3:00:00
#SBATCH --output=deeptools_profile", i, ".out
#SBATCH --error=deeptools_profile", i, ".err

source ~/anaconda3/bin/activate deeptools

date; hostname

computeMatrix reference-point \\
-S /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/bigwig/", bw_files[i]," \\
-R /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/genome/gencode.vM25.annotation.bed \\
--outFileName ", names(bw_files)[i], ".computeMatrix.mat.gz \\
--outFileNameMatrix ", names(bw_files)[i],".computeMatrix.vals.mat.tab \\
--outFileSortedRegions ", names(bw_files)[i],".sorted_regions.txt \\
-a 3000 -b 3000 \\
--referencePoint TSS \\
--numberOfProcessors 4

date
")
  
  write_lines(script, paste0("bin/deeptools_profile/", names(bw_files)[i], ".sh"))
}

submit_script <- c("#!/bin/bash", paste0("sbatch ", names(bw_files), ".sh"))
write_lines(submit_script, "bin/deeptools_profile/submit.sh")

```


```{bash}
# Use deeptools to plot the TSS profile metaplot
source ~/anaconda3/bin/activate deeptools

# Gene bed file: /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/genome/gencode.vM25.annotation.bed
# Test bigWig: /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/bigwig/KOiTg_48_IP_R1.bigWig


#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=deeptools_profile1
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=michael.smallegan@colorado.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=100gb
#SBATCH --time=10:00:00
#SBATCH --output=deeptools_profile1.out
#SBATCH --error=deeptools_profile1.err
computeMatrix reference-point \
--scoreFileName /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/bwa/mergedLibrary/bigwig/KOiTg_48_IP_R1.bigWig \
--regionsFileName /scratch/Shares/rinn/JR/FIRRE_KO_TG_K27me3_CHIPSEQ/results/genome/gencode.vM25.annotation.bed \
--outFileName KOiTg_48_IP_R1.computeMatrix.mat.gz
--outFileNameMatrix KOiTg_48_IP_R1.computeMatrix.vals.mat.tab
--outFileSortedRegions KOiTg_48_IP_R1_sorted_regions \
--upstream 3000 --downstream 3000 \
--referencePoint TSS \
--numberOfProcessors 4



```

```{bash}
plotHeatmap \
-m WT_CTL_48.computeMatrix.mat.gz \
-out WT_CTL_48_TSS_2kb.pdf \
--colorMap jet \
--missingDataColor "#FFF6EB" \
--heatmapHeight 15 \
--plotTitle 'WT_CTL_48_TSS_2kb'



plotHeatmap \
-m WTiTg_48.computeMatrix.mat.gz \
-out WTiTg_48_TSS_2kb.pdf \
--colorMap jet \
--missingDataColor "#FFF6EB" \
--heatmapHeight 15 \
--plotTitle 'WTiTg_48_TSS_2kb'
```

#TODO what is going on here?
```{r}
fl <- list.files("bin/deeptools_profile", 
                 pattern = ".mat.tab",
                 full.names = TRUE)
names(fl) <- gsub("bin/deeptools_profile/|.computeMatrix.vals.mat.tab", "", fl)

meta_df <- tibble(sample = character(),
                  position = integer(),
                  signal = numeric())
library(tidyverse)
for(i in 1:length(fl)) {
  tss_matrix <- read.table(fl[i], sep = "\t", skip = 3, header = F) %>%
    as.matrix()
  meta_profile <- colMeans(tss_matrix, na.rm = T)
  tmp_meta_df <- tibble(sample = names(fl)[[i]],
                        position = 1:length(meta_profile),
                        signal = meta_profile)
  meta_df <- bind_rows(meta_df, tmp_meta_df)
}

write_csv(meta_df, "results/metaplot_df.csv")

meta_df <- meta_df %>%
  mutate(ip = ifelse(grepl("IP", sample), "IP", "control"),
         firre_ko = ifelse(grepl("KO", sample), "KO", "WT"),
         firre_inducible = ifelse(grepl("iTg", sample), "firre_induced", "control"),
         timepoint = ifelse(grepl("48", sample), 48, 0))
ggplot(meta_df %>% filter(ip == "control"), aes(x = position, y = signal, color = timepoint)) +
  geom_point()

unique(meta_df$sample)
# Maybe just try normalizing to the control for each sample will do it?
firre_on_samples <- c("KOiTg_48",
"WTiTg_48",
"WT_CTL",
"WTiTg")

firre_induced_samples <- c("KOiTg_48",
"WTiTg_48")
meta_norm <- meta_df %>%
  mutate(sample = gsub("_IP", "", sample)) %>%
  pivot_wider(names_from = ip, values_from = signal) %>%
  mutate(subtracted = IP - control,
              divided = IP / control) %>%
    mutate(firre_ko = ifelse(grepl("KO", sample), "KO", "WT"),
         firre_inducible = ifelse(grepl("iTg", sample), "firre_induced", "control"),
         timepoint = ifelse(grepl("48", sample), 48, 0),
         firre_on = ifelse(sample %in% firre_on_samples, "firre_on", "firre_off"),
         firre_induced = ifelse(sample %in% firre_induced_samples, "firre_induced", "not_fi")) %>%
  group_by(sample) %>%
  mutate(max_val = max(IP)) %>%
  rowwise() %>%
  mutate(max_norm = IP / max_val)

unique(meta_norm$sample)
ggplot(meta_norm, aes(x = position, y = max_norm, color = firre_on)) +
  geom_point()
ggplot(meta_norm, aes(x = position, y = max_norm, color = firre_induced, group = sample)) +
  geom_line()
unique(meta_df$sample)

```




