---
title: "ATAC-seq"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../../util/_plot_theme.R")
source("../../util/_util.R")
source("../01_setup/assumptions.R")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh)
```

# Laoding in atac seq data from NF_core pipeline
# ATAC seq was performed across the short time course (0 - 2.5 hours every 30 min)

# Atac counts
# 
```{r load, include=FALSE}

# All needed ATACseq files : g2s, tpm, atac_samples, atac_counts, atac_consensus, tac_consensus_gr
load("../01_setup/results/atacseq_data.RData", verbose = T)

# atac_counts is 
nrow(atac_counts)
```

# Runnning Deqseq between atac samples in control line and induced "condition"
# Running Deseq on fold change between timepoints

```{r Deseq on atac peaks between control and induced conditions}

if(!file.exists("results/atacseq_firre_vs_control.RData")) {
  
  
  # here we are comparing all timepoints between control and induced
  # significant peaks will be in induced but not control
  atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, 
                                     colData = atac_samples, 
                                     design = ~ condition)
  # Deseq results
  atac_dds <- DESeq(atac_dds)
  
  # clearning up results
  atac_lfc <- results(atac_dds) %>%
    as.data.frame() %>%
    rownames_to_column("interval_id") %>%
    merge(atac_consensus %>% 
            dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                          Distance.to.TSS, Chr, Start, End)) %>% 
    
    # naming each peak by chromosome start and end
    unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
    unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
  # renaming col 8 & 9
  names(atac_lfc)[8] <- "gene_name"
  names(atac_lfc)[9] <- "gene_id"
  
  # creating shrunken lfc values for peaks differential in control and induced
  # NOTE: we do NOT use shrunken LFC moving forward for thresholding is on lfc
  atac_shrnklfc <- lfcShrink(atac_dds, 
                             coef = "condition_firre_vs_control", 
                             type = "apeglm") %>%
    as.data.frame() %>%
    rownames_to_column("interval_id")
  
  # cleaning up shrnklfc
  atac_shrnklfc <- atac_shrnklfc %>%
    dplyr::select(interval_id, log2FoldChange) %>%
    dplyr::rename(l2fc_shrnk = log2FoldChange)
  
  # merging lfc and shrunken lfc 
  atac_lfc <- atac_lfc %>% left_join(atac_shrnklfc)
  
  # running rlog normalization 
  # this will be used later for plotting atac changes vs RNA changes
  atac_rlog_counts <- rlog(atac_dds, blind = TRUE) %>%
    
    # assay extracts the data from the Deseq object (results)
    assay() %>%
    as.data.frame() %>%
    rownames_to_column("interval_id") %>%
    pivot_longer(2:ncol(.), names_to = "sample_name", values_to = "count") %>%
    left_join(atac_samples)
  
# Deseq on ATAC peaks across time points
# determining fold change across each time point
  fi_atac_samples <- atac_samples %>%
    # just firre induced WT and KO and lacks the control samples
    # we will merge sig peaks here with those that were sig between control and induced
    filter(firre_induced == "firre_induced")
  fi_atac_counts <- atac_counts[,fi_atac_samples$sample_name]
  atac_vszero_dds <- DESeqDataSetFromMatrix(countData = fi_atac_counts, 
                                            colData = fi_atac_samples, 
                                            design = ~ timepoint_minutes)
  # running Deseq on the model above
  atac_vszero_dds <- DESeq(atac_vszero_dds)
  vs_zero_res <- resultsNames(atac_vszero_dds)
  vs_zero_res <- vs_zero_res[grepl("timepoint", vs_zero_res)]
  
  # cleaning up to make lfc on each result file and then binding together
  atac_vszero_lfc <- lapply(vs_zero_res, function(x) {
    results(atac_vszero_dds, name = x) %>%
      as.data.frame() %>%
      rownames_to_column("interval_id") %>% 
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_minutes_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  # Creating shrunken lfc values (shrnklfc)
  atac_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(atac_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("interval_id") %>% 
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_minutes_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  # saving
  save(atac_lfc, atac_vszero_lfc, atac_vszero_shrnklfc, 
       atac_rlog_counts, file = "results/atacseq_firre_vs_control.RData")
}
```

# Loading in the results generated from above
```{r Deseq on ataac peaks between control and induced conditions}

load("results/atacseq_firre_vs_control.RData", verbose = T)
```

# Now we are going to do the same analysis to find peaks sig in Dox control smaples
# We will use KO and WT cntrls as replicates and look for diff peaks per timepoint

```{r peaks sig in dox controls (WT and KO) over time }

# Doxycycline effects.
if(!file.exists("results/dox_atac_lfc.RData")) {
  
  dox_atac_samples <- atac_samples %>% filter(firre_induced == "control")
  dox_atac_counts <- atac_counts[,dox_atac_samples$sample_name]  
  
  # Model is time and control samples only
  dox_atac_dds <- DESeqDataSetFromMatrix(countData = dox_atac_counts, 
                                         colData = dox_atac_samples, 
                                         design = ~ timepoint_minutes)
  # run Deseq
  dox_atac_dds <- DESeq(dox_atac_dds)
  
  dox_atac_lfc <- results(dox_atac_dds) %>%
    as.data.frame() %>%
    rownames_to_column("interval_id") %>%
    merge(atac_consensus %>% 
            dplyr::select(interval_id, Gene.Name, Nearest.PromoterID, 
                          Distance.to.TSS, Chr, Start, End)) %>% 
    # naming peaks uniquely
    unite(ucsc_coord, Chr, Start, remove = FALSE, sep = ":") %>%
    unite(ucsc_coord, ucsc_coord, End, sep = "-", remove = FALSE)
  names(dox_atac_lfc)[8] <- "gene_name"
  names(dox_atac_lfc)[9] <- "gene_id"
  
  save(dox_atac_lfc, file = "results/dox_atac_lfc.RData")
}

load("results/dox_atac_lfc.RData", verbose = T)

# Filtering to significant peaks
dox_sig <- dox_atac_lfc %>% 
  filter(padj < pval_thresh)
nrow(dox_sig)
```
### Result: No peaks are significant in dox control
### REsult: this ideal as we don't have any dox specific effects



# Significant ATAC peaks changing versus 0 timepoint 

```{r sig ATAC peaks over time versus 0 timepoint}

# filtering pval 
atac_dox_sig <- atac_lfc %>%
  filter(padj < pval_thresh)

# crossing with dox controlled peaks (no dox peaks so not doing much)
atac_vszero_sig <- atac_vszero_lfc %>%
  filter(interval_id %in% atac_dox_sig$interval_id)

# Saving sig ATAC peaks versus 0 timepoint
save(atac_vszero_sig, file = "results/atac_vszero_sig.RData")

```

# ATAC overlaps with promoters of short combined time course sig genes
# Look for overlaps of peaks and differentially expressed genes
# We will find overlaps with promoters using the peak center (MACS) for overlaps.

```{r loading in short time course combnined sig genes adn finding overlaps with their promoters}

#TODO go over with Michael
# Loading combined short time course sig genes 
load("../08_combined_wt_ko_short_time_course/results/combined_short_vszero_sig.RData")
     
     # short_vszero_sig.RData
sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]

# making a promoters GenomicRanges with promoters function
promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  
  # taking all genes except processed pseudogenes
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

# finding overlaps of promoter regions in promoters_df
promoters_gr <- promoters_gr[promoters_gr$gene_id %in%promoters_df$gene_id]

# establishing peak centers building Ranges for sig atac peaks
peak_centers <- start(sig_atac_gr) + round(width(sig_atac_gr)/2)
sig_atac_center <- GRanges(seqnames = seqnames(sig_atac_gr),
                           ranges = IRanges(start = peak_centers,
                                            end = peak_centers),
                           interval_id = sig_atac_gr$interval_id)

# setting peak to 100Kb up and downstream from peak center
offset <- 1e5

# creating 10Kb up and downstream of peak center with offset
start(sig_atac_center) <- start(sig_atac_center) - offset
end(sig_atac_center) <- end(sig_atac_center) + offset

# overlaps with TSS for sig atac peak centers
peak_tss_overlaps <- findOverlaps(promoters_gr, sig_atac_center)

# making Granges into data frame
atac_df <- sig_atac_center %>% as.data.frame()
promoters_df <- promoters_gr %>% as.data.frame()

# finding a bunch of overlapping features with atac peak centers 
# @to is atac sig GR @ from is promters GR
overlapping_df <- atac_df[peak_tss_overlaps@to, ]
overlapping_df$gene_id <- promoters_df$gene_id[peak_tss_overlaps@from]
overlapping_df$gene_name <- promoters_df$gene_name[peak_tss_overlaps@from]
overlapping_df$gene_tss <- promoters_df$start[peak_tss_overlaps@from]
overlapping_df$gene_chr <- promoters_df$seqnames[peak_tss_overlaps@from]
overlapping_df$gene_strand <- promoters_df$strand[peak_tss_overlaps@from]

# calculating distance from TSS
# subtracting peak center from the TSS to find distance
overlapping_df <- overlapping_df %>%
  mutate(dist_to_tss = gene_tss - (start+offset),
         dist_to_peak = (start+offset) - gene_tss)

# some distances are positive and negative depending on up or down stream
# fixing to plot so they are all in the same 
# Flip the minus strand differences so that it's a TSS centric view
overlapping_df[overlapping_df$gene_strand == "-", "dist_to_peak"] <- -1* overlapping_df[overlapping_df$gene_strand == "-", "dist_to_peak"] 

# finding overlaps with sig genes in short combined
overlapping_df$deg <- overlapping_df$gene_name %in%short_vszero_sig$gene_name

# making a new col of random values "runif" r uniform dist
overlapping_df$y <- runif(nrow(overlapping_df))

# Gfap has a closer peak changes that explains its expression change.
# there are two peaks near GFAP so grabbing the closest one
overlapping_df$deg[overlapping_df$interval_id == "Interval_26047" &
                     overlapping_df$gene_name == "Gfap"] <- FALSE

# plotting position of ATAC peak distance to nearest TSS
ggplot(overlapping_df %>% filter(gene_name != "Firre", deg == TRUE), aes(x = dist_to_peak, fill = deg)) +
  geom_density(adjust = 1.2) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_rug() +
  scale_fill_manual(values = c("#424242","#a8404c")) +
  scale_x_continuous(breaks = seq(from = -1e5, to = 1e5, by = 5e4),
                     labels = c("-100", "-50", "TSS", "50", "100"),
                     limits = c(-1e5,1e5))

# saving
ggsave("figures/atac_peak_tss_density.pdf", height = 2, width = 2.5)
```

# Significant ATAC-seq differential accessibility volcano plot

```{r sig ATAC peak volcano plot}

atac_lfc <- atac_lfc %>%
  mutate(sig = ifelse(padj < pval_thresh, "sig", "ns"))

# We don't need 50,000 overlapping points at approximately
# subsetting to the non-significant peaks = 50,000
# let's only take 500 ns peaks for volcano plot

atac_lfc_ns <- atac_lfc %>%
  filter(sig == "ns",
         !is.na(padj)) %>%
  sample_n(500)

# merging atac lfc with atac_lfc_ns
atac_lfc_subset <- bind_rows(atac_lfc %>% filter(sig == "sig"),
                             atac_lfc_ns)

# Let's label the number of up and down peaks using if else > 0 lfc
atac_lfc_summary <- atac_lfc %>% filter(sig == "sig") %>%
  mutate(direction = ifelse(log2FoldChange > 0, "up", "down")) %>%
  group_by(direction) %>%
  summarize(count = n())

# plotting
ggplot(atac_lfc_subset, aes(x = log2FoldChange, y = -log10(padj), color = sig)) +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = -log10(pval_thresh), lty = 2) +
  geom_point() +
  annotate("text", x = -1, y = 15, label = paste0("down=", atac_lfc_summary$count[atac_lfc_summary$direction == "down"])) +
  annotate("text", x = 1, y = 15, label = paste0("up=", atac_lfc_summary$count[atac_lfc_summary$direction == "up"])) +
  xlim(-2,2) +
  guides(color = FALSE)

# saving
ggsave("figures/atac_seq_volcano.pdf", height = 2, width = 2, useDingbats = FALSE)
```
### Result: 51 peaks open up and 4 close -- thus Firre increases chromatin accessibility


# ATAC sig peak heat map and Volcano plot
```{r ATAC sig peaks heatmap and Volcano plot}

# creating heatmap matrix
atac_lfc_matrix <- atac_vszero_sig %>%
  dplyr::select(interval_id, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("interval_id") %>%
  as.matrix()

# row centering 
ordering <- atac_vszero_sig %>%
  filter(abs(log2FoldChange) > 0.2) %>%
  group_by(interval_id) %>%
  summarize(first_tp_de = min(timepoint),
            max_fc = max(log2FoldChange)) %>%
  arrange(first_tp_de,
          -max_fc)

# adding 0 timepoint
atac_lfc_matrix <- cbind(matrix(0, nrow = nrow(atac_lfc_matrix), ncol = 1), atac_lfc_matrix)
colnames(atac_lfc_matrix)[[1]] <- "0"

# ordering peaks
atac_lfc_matrix <- atac_lfc_matrix[ordering$interval_id,]

row_ha = rowAnnotation(deg = as.numeric(rownames(atac_lfc_matrix) %in% overlapping_df$interval_id[overlapping_df$deg == TRUE]),
                       col = list(deg = c("1" = "black", "0" = "white")))

# seeing how many peaks are in matrix == 55
table(as.numeric(rownames(atac_lfc_matrix) %in% overlapping_df$interval_id[overlapping_df$deg == TRUE]))

# Plotting
pdf(paste0("figures/atac_peaks_heatmap_", thresh, ".pdf"), 
    width = 4, height = 3.5)
ht1 <- Heatmap(atac_lfc_matrix, 
               name = "l2fc",
               cluster_columns = FALSE, show_row_names = FALSE, 
               cluster_rows = TRUE,
               left_annotation = row_ha,
               col = colorRamp2(seq(-2,2,length.out = 100), col_pal10))
draw(ht1)
dev.off()
draw(ht1)
```

# ATAC-seq vs. gene expression timing
### Here wew will compare the half maximum peak size and tpm of induction of ATAC peaks overlapping induced genes (11) - including Firre

```{r ATACseq vs RNAseq timing of Firre targtes}

load("../01_setup/results/rnaseq_data.RData")

# filtering to significant genes that have atac peaks within 100Kb
peak_gene_overlaps <- overlapping_df %>% filter(deg == TRUE)

# subsetting to interval_id and gene name
peak2gene <- peak_gene_overlaps %>%
  dplyr::select(interval_id, gene_name)

# filter expression values of genes that have sig deg and sig atac
expr <- tpm %>% 
  filter(gene_name %in% peak_gene_overlaps$gene_name) %>%
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  left_join(samples) %>%
  filter(cell_type == "ESC", timecourse_length == "short", firre_induced == "firre_induced") %>%
  group_by(timepoint, gene_name) %>%
  summarize(tpm = mean(tpm)) %>%
  mutate(timepoint = as.numeric(as.character(timepoint))) %>%
  left_join(peak2gene)

# getting rlog counts for atac peaks
peak <- atac_rlog_counts %>%
  filter(interval_id %in% peak_gene_overlaps$interval_id,
         firre_induced == "firre_induced") %>%
  group_by(timepoint, interval_id) %>%
  summarize(atac_peak_count = mean(count)) %>%
  left_join(peak2gene)


# merging in tpm of deg genes and rlog of sig atac peaks overlapping deg
combined_expr_peak <- expr %>% left_join(peak)

# Normalize values between min and max = percentage of signal for each gene
# 0 = min val and 1 = max
# this just defines the function that will be run (range01) on each row 
range01 <- function(x){(x-min(x, na.rm = TRUE))/(max(x, na.rm = TRUE)-min(x, na.rm = TRUE))}

# now running min max normalization function for tpm adn atac counts
combined_expr_peak <- combined_expr_peak %>%
  group_by(gene_name) %>%
  mutate(rel_tpm = range01(tpm),
         rel_atac = range01(atac_peak_count),
         scaled_tpm = scale(tpm),
         scaled_atac = scale(atac_peak_count))

# plotting expression and atac values across time for each gene
ggplot(combined_expr_peak, aes(x = timepoint, y = rel_tpm)) +
  geom_point(color = "#a8404c") +
  geom_line(color = "#a8404c") +
  geom_point(data = combined_expr_peak, aes(x = timepoint, y = rel_atac),
             color = "#424242") +
  geom_line(data = combined_expr_peak, aes(x = timepoint, y = rel_atac),
              color = "#424242") +
  facet_wrap(~gene_name)


# Now we are summing tpm of all genes at each timepoint 
# removing firre 
meta_expr_peak <- combined_expr_peak %>%
  filter(gene_name != "Firre", timepoint <= 150) %>%
  group_by(timepoint) %>%
  summarize(tpm = sum(tpm),
            atac = sum(atac_peak_count)) %>%
  mutate(rel_tpm = range01(tpm),
         rel_atac = range01(atac))

# making a col of atac/rna and another col of atac_values/rna_values
meta_expr_peak_l <- meta_expr_peak %>%
  dplyr::select(timepoint, rel_tpm, rel_atac) %>%
  pivot_longer(2:3, names_to = "signal_type", values_to = "rel_signal") %>%
  mutate(signal_type = gsub("rel_tpm", "expr", signal_type),
         signal_type = gsub("rel_atac", "atac", signal_type)) %>%
  filter(!is.na(rel_signal))


# making a new col weight where 0 and 150 will be strong penalized if lowess is far off
meta_expr_peak_l <- meta_expr_peak_l %>%
  mutate(weight = ifelse(timepoint %in% c(0,150), 100, 1))


# These plots have a roughly linear trend. We can use the linear fit to determine the half-max
atac_meta <- meta_expr_peak_l %>% filter(signal_type == "atac", !is.na(rel_signal))
atac_mod <- loess(timepoint ~ rel_signal, data = atac_meta, span = 1.1, weights = atac_meta$weight)
atac_half_max <- predict(atac_mod, data.frame(rel_signal = 0.5))
expr_meta <- meta_expr_peak_l %>% filter(signal_type == "expr")
expr_mod <- loess(timepoint ~ rel_signal, data = expr_meta, span = 1.1, weights = expr_meta$weight)
expr_half_max <- predict(expr_mod, data.frame(rel_signal = 0.5))

midpoint <- atac_half_max + (expr_half_max - atac_half_max)/2
offset <- expr_half_max - atac_half_max

meta_expr_peak_l$signal_type <- factor(meta_expr_peak_l$signal_type, levels = c("expr", "atac"))

# plotting time dynamics
ggplot(meta_expr_peak_l, aes(y = timepoint, x = rel_signal, color = signal_type)) +
  geom_point() +
  geom_smooth(method = "loess", se = FALSE, span = 1.1, aes(weight = c(100,100,1,1,1,1,1,1,1,1,100,100))) +
  # geom_abline(intercept = 0, slope = 0.0075164) +
  geom_segment(aes(y = atac_half_max, yend = expr_half_max, x = 0.5, xend = 0.5), lty = 2) +
  annotate("text", y = midpoint, x = 0.5, label = round(offset, 2)) +
  geom_hline(yintercept = atac_half_max, lty = 2) +
  geom_hline(yintercept = expr_half_max, lty = 2) +
  coord_flip() +
  scale_y_continuous(breaks = c(0,atac_half_max, expr_half_max, 150), labels = c(0, round(atac_half_max,1), round(expr_half_max, 1), 150)) +
  guides(color = FALSE)

# saving
ggsave("figures/meta_atac_vs_expression_timing.pdf", height = 2, width = 2)
```

# Plotting ATAC and RNAseq heatmpas of signal for each gene that has ATAC overlapping DEG firre targets

```{r ATACseq vs RNAseq timing of Firre targtes}

# pivot longer so each gene has two enteries per time point one for atac and one for rel_tpm
expr_peak_l <- combined_expr_peak %>%
  dplyr::select(timepoint, gene_name, rel_tpm, rel_atac) %>%
  pivot_longer(3:4, names_to = "signal_type", values_to = "rel_signal") %>%
  mutate(signal_type = gsub("rel_tpm", "expr", signal_type),
         signal_type = gsub("rel_atac", "atac", signal_type))

# Plotting individual gene heat maps of atac and rna signal levels at each timepoint
ggplot(expr_peak_l, aes(x = timepoint, y = signal_type, fill = rel_signal)) +
  geom_tile() +
  facet_wrap(~gene_name) +
   scale_fill_gradientn(colors = col_pal10[50:100])+
  scale_x_continuous(breaks = seq(0,300, by = 60),
                     labels = 0:5)

# saving
ggsave("figures/firre_responder_atac_timing_heatmap.pdf", height = 3.5, width = 4)



# Same plot different color scheme

expr_peak_l <- combined_expr_peak %>%
  dplyr::select(timepoint, gene_name, scaled_tpm, scaled_atac) %>%
  pivot_longer(3:4, names_to = "signal_type", values_to = "rel_signal") %>%
  mutate(signal_type = gsub("scaled_tpm", "expr", signal_type),
         signal_type = gsub("scaled_atac", "atac", signal_type))

ggplot(expr_peak_l, aes(x = timepoint, y = signal_type, fill = rel_signal)) +
  geom_tile() +
  facet_wrap(~gene_name) +
   scale_fill_gradientn(colors = col_pal10) +
  scale_x_continuous(breaks = seq(0,300, by = 60),
                     labels = 0:5)

# saving
ggsave("figures/firre_responder_atac_timing_heatmap_zscaled.pdf", height = 3.5, width = 4)

```
### Result: all atac peaks reach half maximum before RNA as perhaps expected
### Result: gene expression is induced ~57 min after atac peak is called


# Determining where ATAC peaks reside relative to TSS of DEG Firre targets vs all other genes
```{r ATAC positions vs all}

#TODO PRDM14 doesn't show up in this plot.

# Making a GRanges of all atac consensus peaks (union of all peaks called in all timepoints)
sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]
length(sig_atac_gr)

# promoters for all genes except psudo genes
promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

# Changing promters GR to not have pseudo genes.
promoters_gr <- promoters_gr[promoters_gr$gene_id %in%promoters_df$gene_id]

table(promoters_df$gene_type)

# defining ATAC peak centers
sig_atac_center <- GRanges(seqnames = seqnames(sig_atac_gr),
                           ranges = IRanges(start = start(sig_atac_gr) + ((end(sig_atac_gr) - start(sig_atac_gr))/2),
                                            end = start(sig_atac_gr) + ((end(sig_atac_gr) - start(sig_atac_gr))/2) + 1),
                           interval_id = sig_atac_gr$interval_id)

# Determinign how far to the right or left a gene can be ( in paper we used 100Kb 1e5 )
offset <- 1e5

# Creating ranges with offset
start(sig_atac_center) <- start(sig_atac_center) - offset
end(sig_atac_center) <- end(sig_atac_center) + offset

# finding the overlaps with promoters and offset
peak_tss_overlaps <- findOverlaps(promoters_gr, sig_atac_center)

# finding a bunch of features of the overlaps
atac_df <- sig_atac_center %>% as.data.frame()
promoters_df <- promoters_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)

# adding features to the atac peaks wrt/ overlapping promoter (@to)
overlapping_df <- atac_df[peak_tss_overlaps@to, ]

# peak_tss_overlaps @ from is promoter centric overlap features
overlapping_df$gene_id <- promoters_df$gene_id[peak_tss_overlaps@from]
overlapping_df$gene_name <- promoters_df$gene_name[peak_tss_overlaps@from]
overlapping_df$gene_tss <- promoters_df$start[peak_tss_overlaps@from]
overlapping_df$gene_chr <- promoters_df$seqnames[peak_tss_overlaps@from]


# creating a col for distance from TSS for all consensus peaks
overlapping_df <- overlapping_df %>%
  mutate(dist_to_tss = gene_tss - (start+offset))

# 55 peaks have overlaps 
length(unique(overlapping_df$interval_id))
overlapping_df$deg <- overlapping_df$gene_name %in%short_vszero_sig$gene_name
overlapping_df$y <- runif(nrow(overlapping_df))

# Gfap has a closer peak changes that explains its expression change.
overlapping_df$deg[overlapping_df$interval_id == "Interval_26047" &
                     overlapping_df$gene_name == "Gfap"] <- FALSE

ggplot(overlapping_df %>% filter(gene_name != "Firre"), aes(x = dist_to_tss, y = y, color = deg)) +
  geom_point() +
  facet_grid(deg~.)

```

# Determining the abundance of max lfc 
### This is using the combined short course versus 0 values
```{r off to on genes after firre induction - short time course}

# laoding in RNAseq data for TPM table of all genes and timepoints
load("../01_setup/results/rnaseq_data.RData", verbose = T)

# filtering to significant firre target genes
sig_gene_tpm <- tpm %>%
  
  #filtering to sig deg in short time course 
  filter(gene_id %in% short_vszero_sig$gene_id) %>%
  # pivot longer so each gene entry will now have all the time point values as rows
  pivot_longer(3:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  # getting JR# conversion
  left_join(samples) %>%
  filter(cell_type == "ESC",firre_induced == "firre_induced", as.numeric(as.character(timepoint)) == 0,
         timecourse_length == "short") %>%
  group_by(gene_id, gene_name) %>%
  # summarizing mean over all timepoints?
  # TODO what is TPM plotted
  summarize(tpm = mean(tpm))

# making a column to record if sig gene has an atac peak
sig_gene_tpm$has_atac_peak <- sig_gene_tpm$gene_name %in% overlapping_df$gene_name[overlapping_df$deg == TRUE]

# removing Firre
sig_gene_tpm$has_atac_peak[sig_gene_tpm$gene_name == "Firre"] <- FALSE

# plotting
ggplot(sig_gene_tpm %>% filter(gene_name != "Firre"), aes(x = tpm, fill = has_atac_peak)) +
  geom_density(alpha = 0.2)
```
### Result: Those Firre targets with an ATAC peak are similar to those with out (mean abundance)
### Result: There is a slight shift of more abundant expression of target gene without atac peak



# Determining which firre target genes have RAP-DNA Overlaps

```{r RAP-DNA overlaps with Firre target genes }

# reading in published RAP peaks from Hacisuleyman NSMB 
pub_rap_peaks <- read.table("data/Significant Firre RAP trans sites copy.txt")

# reading in rap peaks from Loyal's analysis that has more peaks than the more conservative cut off above
loyal_rap_peaks <- read.table("data/mm9_sigUp_FIRRE_RAP_peaks.bed", skip = 1)

# col names read in as V so substituting names
pub_rap_peaks$V1 <- gsub("mm", "chr", pub_rap_peaks$V1)

# creating GRanges of published RAP peaks
pub_rap_peaks_gr <- GRanges(seqnames = pub_rap_peaks$V1,
                            ranges = IRanges(start = pub_rap_peaks$V2,
                                             end = pub_rap_peaks$V3))

# creating GRanges of Loyal's RAP peaks
loyal_rap_peaks_gr <- GRanges(seqnames = loyal_rap_peaks$V1,
                              ranges = IRanges(start = loyal_rap_peaks$V2,
                                               end = loyal_rap_peaks$V3))

# Firre responder GR.
load("../08_combined_wt_ko_short_time_course/results/short_combined_vs_zero.RData", verbose = T)

# loading in sig Firre genes in combined short time course
fr_genes <- genes[genes$gene_id %in% short_vszero_sig$gene_id]
# Extend 50kb on either side
extend_region <- 5e4
orig_widths <- width(fr_genes)
start(fr_genes) <- start(fr_genes) - extend_region
end(fr_genes) <- end(fr_genes) + extend_region
stopifnot(all(width(fr_genes) - orig_widths == extend_region *2))

promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

promoters_gr <- promoters_gr[promoters_gr$gene_id %in% promoters_df$gene_id]
fr_promoters_gr <- promoters_gr[promoters_gr$gene_id %in% short_vszero_sig$gene_id]
orig_widths <- width(fr_promoters_gr)
start(fr_promoters_gr) <- start(fr_promoters_gr) - extend_region
end(fr_promoters_gr) <- end(fr_promoters_gr) + extend_region
stopifnot(all(width(fr_promoters_gr) - orig_widths == extend_region *2))

# Sig ATAC GR
sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]

# Let's look at overlaps 
findOverlaps(pub_rap_peaks_gr, fr_genes)
# NO overlaps

findOverlaps(loyal_rap_peaks_gr, fr_genes)
# 3 overlaps
#TODO are these firre? How to index and find out

# overlaps with promoters
findOverlaps(pub_rap_peaks_gr, fr_promoters_gr)
findOverlaps(loyal_rap_peaks_gr, fr_promoters_gr)
# NO overlaps

# overlaps with atac peaks
findOverlaps(pub_rap_peaks_gr, sig_atac_gr)
findOverlaps(loyal_rap_peaks_gr, sig_atac_gr)
# NO Overlaps

# All Firre responder genes
load("../08_combined_wt_ko_short_time_course/results/short_combined_vs_zero.RData", verbose = T)
load("../04_Firre_ko_long_time_course/results/ko_rescue_long_vszero_sig.RData", verbose = T)
load("../05_Firre_ko_short_time_course/results/ko_rescue_short_vszero_sig.RData", verbose = T)
load("../06_WT_long_time_course/results/wt_overexp_long_vszero_sig.RData", verbose = T)
load("../07_wt_short_time_course/results/wt_overexp_short_vszero_sig.RData", verbose = T)

# combining all sig genes in all time courses
all_firre_responder_genes <- unique(c(short_vszero_sig$gene_id,
                                      ko_rescue_long_vszero_sig$gene_id,
                                      ko_rescue_short_vszero_sig$gene_id,
                                      wt_overexp_long_vszero_sig$gene_id,
                                      wt_overexp_short_vszero_sig$gene_id))

# intersecting to make fr_genes that are sig in all time courses
fr_genes <- genes[genes$gene_id %in% all_firre_responder_genes]

# there are 526 sig genes in all time courses
unique(length(fr_genes))

# Extend 50kb on either side
extend_region <- 5e4

# adding extension to gene annotations
orig_widths <- width(fr_genes)
start(fr_genes) <- start(fr_genes) - extend_region
end(fr_genes) <- end(fr_genes) + extend_region
stopifnot(all(width(fr_genes) - orig_widths == extend_region *2))

# defining promoter regions 
promoters_gr <- promoters(genes, upstream = 1, downstream = 1)
promoters_df <- promoters_gr %>% as.data.frame() %>%
  filter(gene_type %in% c("protein_coding", "processed_transcript", "bidirectional_promoter_lncRNA",
                          "antisense", "lincRNA", "sense_intronic"))

# promoter GR with selected gene types above (not psuedo genes)
promoters_gr <- promoters_gr[promoters_gr$gene_id %in% promoters_df$gene_id]

# promter regions of all sig firre genes
fr_promoters_gr <- promoters_gr[promoters_gr$gene_id %in% all_firre_responder_genes]

length(fr_promoters_gr)
# still have all 526 = good

# extending the annotations by 50Kb
orig_widths <- width(fr_promoters_gr)
start(fr_promoters_gr) <- start(fr_promoters_gr) - extend_region
end(fr_promoters_gr) <- end(fr_promoters_gr) + extend_region
stopifnot(all(width(fr_promoters_gr) - orig_widths == extend_region *2))

# Let's look at overlaps
findOverlaps(pub_rap_peaks_gr, fr_genes)
# No overlaps

findOverlaps(loyal_rap_peaks_gr, fr_genes)
# 71 overlaps 
#TODO how to index 

findOverlaps(pub_rap_peaks_gr, fr_promoters_gr)
# No overlaps

findOverlaps(loyal_rap_peaks_gr, fr_promoters_gr)
# 49 overlaps
```



```{r}

#TODO not sure
# sftp://fiji.colorado.edu/scratch/Shares/rinn/Michael/RAP/archive/FromJesse/Firre-Round4/Firre-6wash_vs_Input.W2000_O1000.p10.r10.bed
# sftp://fiji.colorado.edu/scratch/Shares/rinn/Michael/RAP/firre_rap/results/bwa/mergedLibrary/macs/broadPeak/firre_rap_R1_peaks.broadPeak

orig_rap_peaks <- read.table("/scratch/Shares/rinn/Michael/RAP/archive/FromJesse/Firre-Round4/Firre-6wash_vs_Input.W2000_O1000.p10.r10.bed")
colnames(orig_rap_peaks) <- c("chr", "start", "end")

orig_rap_peaks_gr <- GRanges(seqnames = orig_rap_peaks$chr,
                             ranges = IRanges(start = orig_rap_peaks$start,
                                              end = orig_rap_peaks$end))

sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]

new_rap_peaks <- rtracklayer::import("/scratch/Shares/rinn/Michael/RAP/firre_rap/results/bwa/mergedLibrary/macs/broadPeak/firre_rap_R1_peaks.broadPeak")

findOverlaps(sig_atac_gr, new_rap_peaks)
sig_atac_gr[34]

```














# Seems redundant with above
```{r}

#TODO what to keep !

# Find out which DE genes have an ATAC peak changing nearby
load("../11_short_timecourse_combined/results/short_vszero_sig.RData")

sig_genes_gr <- genes[genes$gene_name %in% short_vszero_sig$gene_name]

sig_genes_df <- sig_genes_gr %>% as.data.frame()
length(sig_genes_gr)
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 10000
end(sig_genes_gr) <- end(sig_genes_gr) + 10000

sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% unique(atac_vszero_sig$interval_id)]
length(sig_atac_gr)
peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)
atac_df <- sig_atac_gr %>% as.data.frame()
sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)
overlapping_df <- atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 10000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]
length(unique(overlapping_df$gene_name))


overlapping_df <- overlapping_df %>%
  mutate(peak_center = start + (end - start),
         dist_to_tss = peak_center - gene_tss)

overlapping_df <- overlapping_df %>%
  mutate(sig_peak = interval_id %in% sig_atac_peaks$interval_id)
```



```{r}
sig_peak_counts <- atac_rlog_counts %>%
  filter(interval_id %in% sig_atac_peaks$interval_id)

sig_peak_count_summary <- sig_peak_counts %>%
  filter(firre_induced == "firre_induced") %>%
  group_by(interval_id, timepoint) %>%
  summarize(count = mean(count))

zero_tp <- sig_peak_count_summary %>%
  filter(timepoint == 0) %>%
  dplyr::select(-timepoint) %>%
  dplyr::rename(zero_val = count)

sig_peak_count_summary <- sig_peak_count_summary %>%
  left_join(zero_tp) %>%
  mutate(fc = count / zero_val,
         l2fc = log2(fc))

sp_matrix <- sig_peak_count_summary %>%
  dplyr::select(interval_id, timepoint, l2fc) %>%
  pivot_wider(names_from = timepoint, names_sort = T, values_from = l2fc) %>%
  column_to_rownames("interval_id") %>%
  as.matrix()

sp_scaled <- t(scale(t(sig_peak_count_summary)))
pheatmap::pheatmap(sp_matrix, cluster_cols = F, show_rownames = F, breaks = seq(-0.25, 0.25, length.out = length(col_pal10)),
                   color = col_pal10)


short_lfc <- short_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()
```



```{r}
ggplot(rlog_df %>% filter(interval_id == "Interval_76872", firre_induced == "firre_induced"), aes(x = timepoint, y = count, color = firre_ko)) +
  geom_point()


summary(mod1)
sig_atac_peaks <- atac_lfc %>%
  filter(padj <= 0.05)

# abs(log2FoldChange) > log2(1.5)
```

```{r}

load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData", verbose = T)
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")

ko_rescue_1 <- unique(ko_rescue_short_vszero_sig$gene_name)
wt_overexp_2 <- unique(wt_overexp_short_vszero_sig$gene_name)
short_3 <- unique(short_vszero_sig$gene_name)

individual_overlaps_12 <- intersect(ko_rescue_1, wt_overexp_2)
all_overlaps_123 <- intersect(individual_overlaps_12, short_3)
all_sig_genes <- unique(c(ko_rescue_1, wt_overexp_2, short_3))
gencode_genes <- rtracklayer::import("../util/gencode.vM25.annotation.genes.gtf")

# sig_atac_gr <- atac_consensus_gr[atac_consensus_gr$interval_id %in% sig_atac_peaks$interval_id]







# What is the difference between the genes changing without an ATAC-seq peak change
# and those that do? 
# Are we over controlling for dox? Are there ATAC-seq windows in these regions that have 
# linearly increasing trends, but are not significant when compared to the dox control? 




# Let's retrieve the peak counts for the ATAC-seq peaks in the region
# of these genes.
peaks_of_interest <- unique(overlapping_df$interval_id)

library(broom)

peak_id <- peaks_of_interest[[1]]
# Run linear model
peak_lm <- function(peak_id) {
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced")
  mod1 <- lm(peak1$count~peak1$timepoint)
  res1 <- tidy(mod1) 
  res1$intercept <- res1$estimate[[1]]
  res1 <- res1[2,2:ncol(res1)]
  res1 <- res1 %>%
    mutate(interval_id = peak_id) %>%
    dplyr::rename(slope = estimate)
  #WT
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced",
                              firre_ko == "WT")
  mod2 <- lm(peak1$count~peak1$timepoint)
  res2 <- tidy(mod2) 
  res2$intercept <- res2$estimate[[1]]
  res2 <- res2[2,2:ncol(res2)]
  res2 <- res2 %>%
    dplyr::select(estimate, p.value, intercept) %>%
    dplyr::rename(WT_slope = estimate,
                  WT_pval = p.value,
                  WT_intercept = intercept)
  #KO
  peak1 <- rlog_df %>% filter(interval_id == peak_id, 
                              firre_induced == "firre_induced",
                              firre_ko == "KO")
  mod3 <- lm(peak1$count~peak1$timepoint)
  res3 <- tidy(mod3) 
  res3$intercept <- res3$estimate[[1]]
  res3 <- res3[2,2:ncol(res3)]
  res3 <- res3 %>%
    dplyr::select(estimate, p.value, intercept) %>%
    dplyr::rename(KO_slope = estimate,
                  KO_pval = p.value,
                  KO_intercept = intercept)
  res <- cbind(res1, res2)
  res <- cbind(res, res3)
  return(res)
  
  
}

peak_lm_res <- lapply(peaks_of_interest, peak_lm) %>%
  bind_rows()
save(peak_lm_res, file = "results/peak_lm_res.RData")


hmm <- overlapping_df %>%
  left_join(peak_lm_res)
max(peak_lm_res$p.value)
max(hmm$p.value)
hmm$padj <- p.adjust(hmm$p.value)

huh <- hmm %>% filter(p.value < 0.05)
```

# Gene expresion changes vs atac-seq

```{r}
# Let's do rapgef4
huh1 <- hmm %>% filter(gene_name %in% short_3)
goi_atac_peaks <-huh1 %>% filter(gene_name == "Duox1") %>%
  mutate(sig = p.value < 0.05)
unique(goi_atac_peaks$interval_id)
# goi_atac_slopes <- goi_atac_peaks %>%
#   dplyr::select(interval_id, dist_to_tss, slope, p.value, )
huh1_batch <- short_3[1:10]
ggplot(huh1, aes(x = dist_to_tss, y = KO_slope, color = sig_peak)) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point() +
  facet_wrap(~gene_name, scales = "free_x")
ggsave("figures/atacpeak_locations_ko.pdf", height = 7, width = 7)
ggplot(goi_atac_slopes, aes(x = slope, y = -log10(p.value))) +
  geom_point()

poi <- goi_atac_peaks %>% arrange(p.value) %>%
  dplyr::slice(1) %>%
  pull(interval_id)

poi_vals <- rlog_df %>% filter(interval_id == poi, 
                               firre_induced == "firre_induced")

ggplot(poi_vals, aes(x = timepoint, y = count)) +
  geom_point()

poi_mean <- poi_vals %>%
  group_by(timepoint) %>%
  summarize(count = mean(count))
ggplot(poi_mean, aes(x = timepoint, y = count)) +
  geom_point()
ggplot(short_vszero_sig %>% filter(gene_name == "Shf"), 
       aes(x = timepoint, y = log2FoldChange)) +
  geom_point()
range01 <- function(x){(x-min(x))/(max(x)-min(x))}
goi_vals <- short_vszero_sig %>% filter(gene_name == "Shf")
goi_vals$relexp <- range01(goi_vals$log2FoldChange)

ggplot(goi_vals, 
       aes(x = timepoint, y = relexp)) +
  geom_point()
poi_mean$relexp <- range01(poi_mean$count)
ggplot(poi_mean, 
       aes(x = timepoint, y = relexp)) +
  geom_point()


# Combine
combined_vals <- bind_rows(poi_mean %>% dplyr::select(timepoint, relexp) %>%
                             mutate(type = "atac_signal"),
                           goi_vals %>% dplyr::select(timepoint, relexp) %>%
                             mutate(type = "gene_expression"))

ggplot(combined_vals, aes(x = timepoint, y = relexp, color = type, group = type)) +
  geom_point() +
  geom_smooth(se = FALSE, span = 1.1)

ggplot(combined_vals, aes(x = timepoint, y = type, fill = relexp)) +
  geom_tile() +
  scale_fill_gradientn(colors = col_pal10[50:100])

```



```{r}
table(huh$gene_name)
length(unique(huh$gene_name))

table(short_3 %in% unique(huh$gene_name))


huh2 <- hmm %>% filter(WT_pval < 0.05)
table(wt_overexp_2 %in% huh2$gene_name)
table(all_overlaps_123 %in% huh2$gene_name)

huh3 <- hmm %>% filter(KO_pval < 0.05)
table(ko_rescue_1 %in% huh3$gene_name)
table(all_overlaps_123 %in% huh3$gene_name)


table(all_overlaps_123 %in% unique(huh$gene_name))
sig_ov_peaks <- hmm %>% filter(padj < 0.05)

short_vszero_sig <- short_vszero_sig %>%
  mutate(has_atac_peak = gene_id %in% overlapping_df$gene_id)

ggplot(short_vszero_sig, aes(x = timepoint, y= log2FoldChange, color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
ggplot(short_vszero_sig, aes(x = log2FoldChange, y= -log10(padj), color = has_atac_peak)) +
  geom_point() +
  facet_wrap(~has_atac_peak)
load("../01_setup/results/rnaseq_data.RData")
```

```{r}
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")
ko_rescue_short_sig <- unique(ko_rescue_short_vszero_sig$gene_id)
load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData")
wt_overexp_short_sig <- unique(wt_overexp_short_vszero_sig$gene_id)

sig_genes <- intersect(wt_overexp_short_sig, ko_rescue_short_sig)

sig_genes_gr <- gencode_genes[gencode_genes$gene_id %in% sig_genes]
# Extend 10kb
start(sig_genes_gr) <- start(sig_genes_gr) - 50000
end(sig_genes_gr) <- end(sig_genes_gr) + 50000

peak_gene_overlaps <- findOverlaps(sig_genes_gr, sig_atac_gr)
sig_atac_df <- sig_atac_gr %>% as.data.frame()

sig_genes_df <- sig_genes_gr %>% as.data.frame() %>% dplyr::select(gene_name, gene_id, seqnames, start, end)

overlapping_df <- sig_atac_df[peak_gene_overlaps@to, ]
overlapping_df$gene_id <- sig_genes_df$gene_id[peak_gene_overlaps@from]
overlapping_df$gene_name <- sig_genes_df$gene_name[peak_gene_overlaps@from]
overlapping_df$gene_tss <- sig_genes_df$start[peak_gene_overlaps@from] + 50000
overlapping_df$gene_chr <- sig_genes_df$seqnames[peak_gene_overlaps@from]
```




