---
title: "Short timecourse combined"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE,
        dplyr.summarise.inform = FALSE,
        tidyverse.quiet = TRUE)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(seriation)
source("../../util/_plot_theme.R")
source("../../util/_util.R")
source("../01_setup/assumptions.R")
source("../../util/figures.R")

# Keep track of thresholds for figure naming
thresh <- paste0("pval", pval_thresh, "_l2fc", round(l2fc_thresh, 2))
```

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData", verbose = T)
```

# Here we are combining the WT and KO Firre transgene lines in a model
# That will account for both genotype, dox and timepoint
### results will be saved as "short_combined_dox_controlled.RData"

```{r long_timecourse}
if(!file.exists("results/short.RData")) {
  
  # Filter to short timecourse both WT and KO
  short_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short")
  
  # getting salmoon counts for Deseeq
  short_counts <- salmon_gene_counts[,short_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(short_samples) == colnames(short_counts)))
  stopifnot(all(rownames(short_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  # LRT for dox control performed here
  short_dds <- DESeqDataSetFromMatrix(countData = short_counts, 
                                      colData = short_samples, 
                                      design = ~ firre_ko + firre_induced + timepoint + timepoint*firre_induced)
  short_dds <- DESeq(short_dds, test = "LRT", reduced = ~ firre_ko + firre_induced + timepoint)
  
  #TODO got this error
  # 1 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT
  
  # Compile results
  res_names <- resultsNames(short_dds)
  
  # grabbing just the Deseq results we want
  dynamic_res <- res_names[grepl("firre_inducedfirre_induced.timepoint", res_names)]
  
  # cleaning up each result file
  short_lfc <- lapply(dynamic_res, function(x) {
    results(short_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Shrunken LFC results
  short_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(short_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  short_maxfc <- short_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  short_shrnklfc <- short_shrnklfc %>%
    left_join(short_maxfc)
  
  # saving
  save(short_lfc, short_shrnklfc, file = "results/short_combined_dox_controlled.RData")
}

load("results/short_combined_dox_controlled.RData")
```

# Combined WT & KO  Fold changes vs zero timepoint
# Deseq model does not have dox ontrol line in it 

```{r Deseq combined WT and KO short time coruse versus 0 with out control line in model}

if(!file.exists("results/short_vs_zero.RData")) {
  
  # calling out samples for short WT and KO RNAseq
  short_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_induced == "firre_induced")
  
  # getting salmon counts for Deseq
  short_vszero_counts <- salmon_gene_counts[,short_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(short_vszero_samples) == colnames(short_vszero_counts)))
  stopifnot(all(rownames(short_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- model for both genotypes and time but not conrol line LRT test
  # NO LRT test 
  short_vszero_dds <- DESeqDataSetFromMatrix(countData = short_vszero_counts,
                                             colData = short_vszero_samples,
                                             design = ~ firre_ko + timepoint)
  short_vszero_dds <- DESeq(short_vszero_dds)
  res_names <- resultsNames(short_vszero_dds)
  
  
# cleaning up result files & converting to shrnkLFC 

  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  short_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(short_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
 
  # saving
  save(short_vszero_shrnklfc, file = "results/short_combined_vs_zero.RData")
}

load("results/short_combined_vs_zero.RData", verbose = T)

```

# Here we are going to perform Deseq on the combined models versus 0 time point for all timepoints long and short

```{r combined WT and KO all timepoints no control line }
if(!file.exists("results/combined_vs_zero.RData")) {
  
  combined_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           firre_induced == "firre_induced")
  combined_vszero_counts <- salmon_gene_counts[,combined_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(combined_vszero_samples) == colnames(combined_vszero_counts)))
  stopifnot(all(rownames(combined_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- model across all time points versus 0 timepoint
  # NO LRT for dox controls 
  combined_vszero_dds <- DESeqDataSetFromMatrix(countData = combined_vszero_counts,
                                             colData = combined_vszero_samples,
                                             design = ~ firre_ko + timepoint)
  combined_vszero_dds <- DESeq(combined_vszero_dds)
  res_names <- resultsNames(combined_vszero_dds)
  
  # getting result files
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  
  # creating shrunken lfc 
  combined_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(combined_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
  
  # saving 
  save(combined_vszero_shrnklfc, file = "results/combined_vs_zero_all_time_points.RData")
}
```

# Creating list of combined WT and KO short timecourse sig genes
# Will merge versus zero with dox controled and add dox to object name
# We will use Pval and shrnklfc thresholds 

```{r creating sig genes in combined WT and KO short time course}
#TODO start here
short_dox_sig <- short_shrnklfc %>% 
  filter(padj <= pval_thresh)

short_vszero_sig <- short_vszero_shrnklfc %>%
  filter(gene_id %in% short_dox_sig$gene_id)

short_vszero_maxfc <- short_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

short_vszero_sig <- short_vszero_sig %>%
  left_join(short_vszero_maxfc)

short_vszero_sig <- short_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

save(short_vszero_sig, file = "results/combined_short_vszero_sig.RData")
```

```{r}
load("results/short_vszero_sig.RData", verbose = T)
short_vszero_sig <- short_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)
make_timecourse_lfc_plot(deseq_res = short_vszero_sig,
                         genes_to_include = unique(short_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/short_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)
```

```{r}
combined_sig_genes <- unique(short_vszero_sig$gene_id)
rescued_gene_ids_20[rescued_gene_ids_20 %in% combined_sig_genes]
```


```{r}
load("results/short_vszero_sig.RData", verbose = T)
load("results/combined_vs_zero_notimecourselengthfactor.RData", verbose = T)
sig_combined_vs_zero <- combined_vszero_shrnklfc %>% filter(gene_id %in% unique(short_vszero_sig$gene_id))
deseq_res <- sig_combined_vs_zero
genes_to_include <- unique(sig_combined_vs_zero$gene_id)
# Let's plot the downregulated genes separately
 deg <- deseq_res %>%
    filter(gene_id %in% genes_to_include) %>%
    dplyr::select(gene_id, gene_name, baseMean, log2FoldChange, timepoint)
  
  # Add a zero timepoint
  deg_zero <- deg %>%
    dplyr::select(gene_id, gene_name, baseMean) %>%
    distinct() %>%
    mutate(log2FoldChange = 0,
           timepoint = 0) 
  
  deg <- deg %>%
    bind_rows(deg_zero)
  
  # if(length(unique(deg$gene_id)) > 1) {
  #   deg <- deg %>%
  #     mutate(x_adjust = scale(log10(baseMean)),
  #            tp_adjust = timepoint + x_adjust*x_adjust_scale)
  # } else {
  #   deg <- deg %>%
  #     mutate(tp_adjust = timepoint)
  # }
  
  num_deg <- deg  %>%
    group_by(gene_id) %>%
    summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
    mutate(direction = ifelse(max_fc > 0, "up", "down")) %>%
    group_by(direction) %>%
    summarize(num_deg = length(unique(gene_id)))
  
  label_x <- max(deg$timepoint / 2)
  
  # Let's sort on the order of fold-change
  deg_fc <- deg %>%
    group_by(gene_name) %>%
    summarize(max_fc = log2FoldChange[which.max(abs(log2FoldChange))]) %>%
    arrange(-max_fc)
  
  deg$gene_name <- factor(deg$gene_name, levels = deg_fc$gene_name)
  
  g1 <- ggplot(deg, aes(x = timepoint, y = log2FoldChange, group = gene_id)) +
    geom_hline(yintercept = 0) +
    geom_line(alpha = 0.2, color = "gray70") +
    geom_point(alpha = 0.4) +
    facet_wrap(~gene_name, scales = "free_y") +
    scale_x_continuous(breaks = c(0,330, 720, 1440, 2880, 5670), labels = c("0", "5.5", "12","24", "48","96") )
  show(g1)
ggsave("figures/short_and_long_individual_fc_profiles.pdf",
       height = 4.5, width = 6.5)
```

```{r}
# Figure out whether these genes stay differentially expressed
# or return to baseline in the long timecourse.
```


```{r}
load("results/short_vszero_sig.RData", verbose = T)
load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData", verbose = T)
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData")

ko_rescue_1 <- unique(ko_rescue_short_vszero_sig$gene_name)
wt_overexp_2 <- unique(wt_overexp_short_vszero_sig$gene_name)
short_3 <- unique(short_vszero_sig$gene_name)

individual_overlaps_12 <- intersect(ko_rescue_1, wt_overexp_2)
length(individual_overlaps)

length(ko_rescue) - length(individual_overlaps)
length(wt_overexp) - length(individual_overlaps)

all_overlaps_123 <- intersect(individual_overlaps_12, short_3)
ov_23 <- intersect(wt_overexp_2, short_3)
ov_13 <- intersect(ko_rescue_1, short_3)
library(VennDiagram)


draw.triple.venn(area1 = length(ko_rescue_1), 
                 area2 = length(wt_overexp_2), 
                 area3 = length(short_3),
                 n12 = length(individual_overlaps_12),
                 n23 = length(ov_23), 
                 n13 = length(ov_13),
                 n123 = length(all_overlaps_123))

library(eulerr)

set.seed(1)
mat <- cbind(
  A = sample(c(TRUE, TRUE, FALSE), 50, TRUE),
  B = sample(c(TRUE, FALSE), 50, TRUE),
  C = sample(c(TRUE, FALSE, FALSE, FALSE), 50, TRUE)
)
all_genes <- unique(c(ko_rescue_1, wt_overexp_2, short_3))
ov_mat <- cbind(
  rescue = all_genes %in% ko_rescue_1,
  overexpr = all_genes %in% wt_overexp_2,
  combined = all_genes %in% short_3
)
fit2 <- euler(ov_mat)
pdf("figures/short_timecourse_overlap_venn.pdf")
plot(fit2, quantities = TRUE)
dev.off()
```


### Short timecourse Firre responders heatmap

```{r fig.width=4, fig.height=4}
# Let's look at the set of genes that overlap in the two genetic backgrounds
load("../06_short_timecourse_rescue/results/ko_rescue_short_vszero_sig.RData", verbose = T)
load("../09_short_timecourse_overexpression/results/wt_overexp_short_vszero_sig.RData", verbose = T)

overlapping_genes <- unique(ko_rescue_short_vszero_sig$gene_id)[unique(ko_rescue_short_vszero_sig$gene_id) %in% unique(wt_overexp_short_vszero_sig$gene_id)]

# Heatmap of fold-changes for DEGs in the rescue
# Check that there are no duplicate row names.
stopifnot(all(length(unique(short_vszero_sig$gene_id)) == length(unique(short_vszero_sig$gene_name))))

short_lfc <- short_vszero_sig %>%
  dplyr::select(gene_name, timepoint, log2FoldChange) %>%
  pivot_wider(names_from = timepoint, names_sort = TRUE, values_from = log2FoldChange) %>%
  column_to_rownames("gene_name") %>%
  as.matrix()


ordering <- short_vszero_sig %>%
  filter(abs(log2FoldChange) > 0.2) %>%
  group_by(gene_name) %>%
  summarize(first_tp_de = min(timepoint),
            max_fc = max(log2FoldChange)) %>%
  arrange(first_tp_de,
          -max_fc)

# Add a zero column.
short_lfc <- cbind(matrix(0, nrow = nrow(short_lfc), ncol = 1), short_lfc)
colnames(short_lfc)[[1]] <- "0"

short_lfc <- short_lfc[ordering$gene_name,]

pdf(paste0("figures/combined_responders_heatmap_", thresh, ".pdf"), 
    width = 4, height = 3.5)
ht1 <- Heatmap(short_lfc, 
               name = "l2fc",
               cluster_columns = FALSE, show_row_names = TRUE, 
               cluster_rows = FALSE,
               col = colorRamp2(seq(-4,4,length.out = 100), col_pal10))
draw(ht1)
dev.off()
draw(ht1)
```