---
title: "DOX_GENES"
author: "JR"
date: "9/10/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r load, include=FALSE}
load("../01_setup/results/rnaseq_data.RData", verbose = T)
```

# Here we are combining the WT and KO Firre transgene lines in a model
# That will account for both genotype, dox and timepoint
### results will be saved as "short_combined_dox_controlled.RData"

```{r short_full_model_timecourse}
if(!file.exists("results/short.RData")) {
  
  # Filter to short timecourse both WT and KO
  dox_short_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_induced == "control")
  
  # getting salmoon counts for Deseeq
  dox_short_counts <- salmon_gene_counts[,dox_short_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(dox_short_samples) == colnames(dox_short_counts)))
  stopifnot(all(rownames(dox_short_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  # LRT for dox control performed here
  dox_short_dds <- DESeqDataSetFromMatrix(countData = dox_short_counts, 
                                      colData = dox_short_samples, 
                                      design = ~ firre_ko + timepoint + timepoint*firre_ko)
  dox_short_dds <- DESeq(dox_short_dds, test = "LRT", reduced = ~ firre_ko + timepoint)
  
  
  # Compile results
  res_names <- resultsNames(dox_short_dds)
  
  # grabbing just the Deseq results we want
  dynamic_res <- res_names[grepl(".timepoint", res_names)]
  
  # cleaning up each result file
  short_lfc <- lapply(dynamic_res, function(x) {
    results(dox_short_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Shrunken LFC results
  short_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(dox_short_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  short_maxfc <- short_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  short_shrnklfc <- short_shrnklfc %>%
    left_join(short_maxfc)
  
  # saving
  save(short_lfc, short_shrnklfc, file = "results/short_combined_dox_controlled.RData")
}
```

# Dox sig genes short combined
```{r}
short_dox_sig <- short_shrnklfc %>% 
  filter(padj <= pval_thresh)


# determining max fold change 
short_dox_sig <- short_dox_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

# filtering on threshold
short_dox_sig <- short_dox_sig %>%
  filter(max_fc > l2fc_thresh)

short_dox_sig <- merge(short_dox_sig, g2s)
```


# Getting data back out by doing DEseq on control lines only

```{r}
 # calling out samples for short WT and KO RNAseq
  short_vszero_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_induced == "control")
  
  # getting salmon counts for Deseq
  short_vszero_counts <- salmon_gene_counts[,short_vszero_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(short_vszero_samples) == colnames(short_vszero_counts)))
  stopifnot(all(rownames(short_vszero_counts) == genes$gene_id))
  
  # DESeq2 -- model for both genotypes and time but not conrol line LRT test
  # NO LRT test 
  short_vszero_dds <- DESeqDataSetFromMatrix(countData = short_vszero_counts,
                                             colData = short_vszero_samples,
                                             design = ~ firre_ko + timepoint)
  # run careful :)
  short_vszero_dds <- DESeq(short_vszero_dds)
  
  # results
  res_names <- resultsNames(short_vszero_dds)
  
  
# cleaning up result files & converting to shrnkLFC 

  # grabbing relevenat results
  vs_zero_res <- res_names[grepl("_vs_0", res_names)]
  
  # applying shrnkl
  short_vszero_shrnklfc <- lapply(vs_zero_res, function(x) {
    lfcShrink(short_vszero_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("timepoint_|_vs_0", "", result_name)))
  }) %>% bind_rows()
 
  # saving
  save(short_vszero_shrnklfc, file = "results/dox_short_combined_vs_zero.RData")


load("results/short_combined_vs_zero.RData", verbose = T)

```



# We will use Pval and shrnklfc thresholds 

```{r creating sig genes in combined WT and KO short time course}

# filtering to sig genes in dox control LRT test "short_shrnklfc"
# This means induced in WT and KO firre induced, but not in dox control lines
short_dox_sig <- short_shrnklfc %>% 
  filter(padj <= pval_thresh)

# Those genes that were in the combined short versus 0 sig genes
# Those genes that are also in the dox control line
short_vszero_sig <- short_vszero_shrnklfc %>%
  filter(gene_id %in% short_dox_sig$gene_id)

# determining max fold change 
short_vszero_maxfc <- short_vszero_sig %>%
  group_by(gene_id) %>%
  summarize(max_fc = max(abs(log2FoldChange))) 

# merging in new col for max fold change with short_vszero_sig
short_vszero_sig <- short_vszero_sig %>%
  left_join(short_vszero_maxfc)

# filtering on threshold
short_vszero_sig <- short_vszero_sig %>%
  filter(max_fc > l2fc_thresh)

# saving results
save(short_vszero_sig, file = "results/dox_combined_short_vszero_sig.RData")
```


# Plotting 

```{r}
 # changing lfc col name
short_vszero_sig <- short_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)

# Plotting
make_timecourse_lfc_plot(deseq_res = short_vszero_sig,
                         genes_to_include = unique(short_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/dox_short_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)

dox_genes_combined_short <-unique(short_vszero_sig$gene_name)



```


# Short Dox KO genes (now doing seperate in each background)


```{r}
 # Filter to short timecourse both WT and KO
  ko_dox_short_samples <- samples %>%
    filter(cell_type == "ESC",
           timecourse_length == "short",
           firre_induced == "control",
           firre_ko == "KO")

# TODO why are all the time points here ?
# Because it's a factor in the larger data set
ko_dox_short_samples$timepoint

  
  # getting salmoon counts for Deseeq
  ko_dox_short_counts <- salmon_gene_counts[,ko_dox_short_samples$sample_id]
  
  # Check ordering
  stopifnot(all(rownames(ko_dox_short_samples) == colnames(ko_dox_short_counts)))
  stopifnot(all(rownames(ko_dox_short_counts) == genes$gene_id))
  
  # DESeq2 -- controlling for doxycycline; likelihood ratio test
  # LRT for dox control performed here
  ko_dox_short_dds <- DESeqDataSetFromMatrix(countData = ko_dox_short_counts, 
                                      colData = ko_dox_short_samples, 
                                      design = ~ timepoint)
  
  # running Deseq LRT
  ko_dox_short_dds <- DESeq(ko_dox_short_dds)
  
  #TODO no result names

  # Compile results
  res_names <- resultsNames(ko_dox_short_dds)
  
  # grabbing just the Deseq results we want
  dynamic_res <- res_names[grepl("timepoint", res_names)]
  
  # cleaning up each result file
  ko_dox_short_lfc <- lapply(dynamic_res, function(x) {
    results(ko_dox_short_dds, 
            name = x) %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Shrunken LFC results
  ### WARNING RUN TIME LONG ####
  ko_dox_short_shrnklfc <- lapply(dynamic_res, function(x) {
    lfcShrink(ko_dox_short_dds, 
              coef = x,
              type = "apeglm") %>%
      as.data.frame() %>%
      rownames_to_column("gene_id") %>% 
      merge(g2s) %>%
      mutate(result_name = x,
             timepoint = as.numeric(gsub("firre_inducedfirre_induced.timepoint", "", result_name)))
  }) %>% bind_rows()
  
  # Calculate the maximum fold-change in any one timepoint
  ko_dox_short_maxfc <- ko_dox_short_shrnklfc %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  ko_dox_short_shrnklfc <- ko_dox_short_shrnklfc %>%
    left_join(ko_dox_short_maxfc)
  
  # saving
  save(ko_dox_short_lfc, ko_dox_short_shrnklfc, file = "results/ko_dox_short.RData")



# TODO how to make heat map 
  # guessing [] index into counts of sig genes - make matrix - scale - heatmap

```

# Plotting ko dox genes

```{r}
 # changing lfc col name
ko_dox_short_vszero_sig <- short_vszero_sig %>%
  mutate(l2fc_shrunken = log2FoldChange)

# Plotting
make_timecourse_lfc_plot(deseq_res = short_vszero_sig,
                         genes_to_include = unique(ko_dox_short_vszero_sig$gene_id),
                         y_lim = c(-6,6),
                         draw_plot = FALSE)

ggsave(paste0("figures/ko_dox_short_lfc_", thresh, ".pdf"),
       height = 3, width = 3.5, useDingbats = FALSE)

ko_dox_short <-unique(ko_dox_short_vszero_sig$gene_name)



```



